<!doctype html><html lang=en dir=ltr><head><meta charset=utf-8><meta name=viewport content='width=device-width,initial-scale=1'><meta name=description content="The Protein Data Bank (PDB) serves as a central resource for experimentally-determined structures of proteins, nucleic acids, and complex assemblies. All the data housed in the PDB is obtained through experimental techniques such as X-ray crystallography, spectroscopy, NMR, and other methods. This article outlines a methodology for extracting, filtering, and cleaning data from the PDB, ultimately facilitating the kind of analysis detailed in the article Occurrence of protein disulfide bonds in different domains of life: a comparison of proteins from the Protein Data Bank, featured in Protein Engineering, Design and Selection, Volume 27, Issue 3, 1 March 2014, pp."><title>Creating a bioinformatics database for studying disulfide bonds</title>
<link rel=canonical href=https://Nexus-Security.github.io/creating-a-bioinformatics-database-for-studying-disulfide-bonds/><link rel=stylesheet href=/scss/style.min.0304c6baf04e01a8fe70693791cb744d56a3578a3120a8796cefc66825aa39c7.css><meta property='og:title' content="Creating a bioinformatics database for studying disulfide bonds"><meta property='og:description' content="The Protein Data Bank (PDB) serves as a central resource for experimentally-determined structures of proteins, nucleic acids, and complex assemblies. All the data housed in the PDB is obtained through experimental techniques such as X-ray crystallography, spectroscopy, NMR, and other methods. This article outlines a methodology for extracting, filtering, and cleaning data from the PDB, ultimately facilitating the kind of analysis detailed in the article Occurrence of protein disulfide bonds in different domains of life: a comparison of proteins from the Protein Data Bank, featured in Protein Engineering, Design and Selection, Volume 27, Issue 3, 1 March 2014, pp."><meta property='og:url' content='https://Nexus-Security.github.io/creating-a-bioinformatics-database-for-studying-disulfide-bonds/'><meta property='og:site_name' content='Nexus Security'><meta property='og:type' content='article'><meta property='article:section' content='Post'><meta property='article:published_time' content='2024-02-26T00:00:00+00:00'><meta property='article:modified_time' content='2024-02-26T00:00:00+00:00'><meta name=twitter:title content="Creating a bioinformatics database for studying disulfide bonds"><meta name=twitter:description content="The Protein Data Bank (PDB) serves as a central resource for experimentally-determined structures of proteins, nucleic acids, and complex assemblies. All the data housed in the PDB is obtained through experimental techniques such as X-ray crystallography, spectroscopy, NMR, and other methods. This article outlines a methodology for extracting, filtering, and cleaning data from the PDB, ultimately facilitating the kind of analysis detailed in the article Occurrence of protein disulfide bonds in different domains of life: a comparison of proteins from the Protein Data Bank, featured in Protein Engineering, Design and Selection, Volume 27, Issue 3, 1 March 2014, pp."><link rel="shortcut icon" href=/fav.ico></head><body class=article-page><script>(function(){const e="StackColorScheme";localStorage.getItem(e)||localStorage.setItem(e,"light")})()</script><script>(function(){const t="StackColorScheme",e=localStorage.getItem(t),n=window.matchMedia("(prefers-color-scheme: dark)").matches===!0;e=="dark"||e==="auto"&&n?document.documentElement.dataset.scheme="dark":document.documentElement.dataset.scheme="light"})()</script><div class="container main-container flex on-phone--column extended"><aside class="sidebar left-sidebar sticky"><button class="hamburger hamburger--spin" type=button id=toggle-menu aria-label="Toggle Menu">
<span class=hamburger-box><span class=hamburger-inner></span></span></button><ol class=menu-social><li><a href=https://github.com/Nexus-Security target=_blank title=GitHub rel=me><svg xmlns:xlink="http://www.w3.org/1999/xlink" width="150" height="150" viewBox="0 0 150 150"><defs><filter id="alpha" filterUnits="objectBoundingBox" x="0" y="0" width="100%" height="100%"><feColorMatrix type="matrix" in="SourceGraphic" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 1 0"/></filter><mask id="mask0"><g filter="url(#alpha)"><rect x="0" y="0" width="150" height="150" style="fill:#000;fill-opacity:.2;stroke:none"/></g></mask><clipPath id="clip1"><rect x="0" y="0" width="150" height="150"/></clipPath><g id="surface5" clip-path="url(#clip1)"><path style="stroke:none;fill-rule:nonzero;fill:#fff;fill-opacity:1" d="M10.9375 83.867188c0-12.109376 3.46875-21.816407 10.925781-30.535157.71875-.875.925781000000001-2.058593.550781000000001-3.125C21.6875 48.125 21.40625 33.070312 26.488281 17.675781c4.78125 1.207031 16.300781 5.050781 31.65625 16.398438C58.9375 34.65625 59.957031 34.84375 60.894531 34.554688 64.824219 33.382812 71.851562 32.6875 79.6875 32.6875S94.550781 33.386719 98.476562 34.554688C99.417969 34.835938 100.433594 34.65625 101.226562 34.074219 113.96875 24.65625 124.011719 20.445312 129.8125 18.582031c-.476562000000001-1.882812-1.019531-3.746093-1.625-5.589843-4.78125 1.207031-16.300781 5.050781-31.648438 16.394531C95.746094 29.96875 94.730469 30.144531 93.789062 29.867188c-3.925781-1.171876-10.953124-1.875-18.789062-1.875-7.832031.0-14.855469.703124000000003-18.792969 1.871093C55.265625 30.144531 54.246094 29.964844 53.457031 29.382812c-15.355469-11.34375-26.875-15.1875-31.65625-16.402343C16.71875 28.382812 17 43.4375 17.726562 45.523438 18.101562 46.589844 17.890625 47.777344 17.175781 48.648438 9.71875 57.367188 6.25 67.070312 6.25 79.179688c0 23.515624 7.175781 37.921874 17.417969 46.550781C16.007812 116.851562 10.9375 103.582031 10.9375 83.867188zm0 0"/></g><mask id="mask1"><g filter="url(#alpha)"><rect x="0" y="0" width="150" height="150" style="fill:#000;fill-opacity:.101961;stroke:none"/></g></mask><clipPath id="clip2"><rect x="0" y="0" width="150" height="150"/></clipPath><g id="surface8" clip-path="url(#clip2)"><path style="stroke:none;fill-rule:nonzero;fill:#000;fill-opacity:1" d="M32.34375 78.5c3.945312-4 9.382812-5.074219 16.039062-5.074219C51.925781 73.425781 55.824219 73.730469 60.023438 74.054688c4.71875.375 9.589843.71875 14.84375.746093000000002C80.335938 74.917969 85.210938 74.425781 89.976562 74.054688 102.070312 73.125 111.605469 72.367188 117.648438 78.492188c4.527343 4.601562 6.890624 10.328124 7.21875 16.945312C124.925781 94.398438 125 93.382812 125 92.261719c0-7.285157-2.476562-13.492188-7.351562-18.449219-6.042969-6.117188-15.578126-5.367188-27.671876-4.4375C85.210938 69.742188 80.34375 70.238281 74.867188 70.117188 69.914062 70.0625 64.960938 69.8125 60.023438 69.375 55.824219 69.050781 51.929688 68.742188 48.382812 68.742188c-6.65625.0-12.09375 1.070312-16.039062 5.078124C27.476562 78.769531 25 84.976562 25 92.261719 25 93.375 25.074219 94.398438 25.132812 95.445312 25.460938 88.820312 27.820312 83.101562 32.34375 78.5zm0 0"/></g><linearGradient id="linear0" gradientUnits="userSpaceOnUse" x1="10.768" y1="13.86" x2="18.633" y2="21.724" gradientTransform="matrix(6.25,0,0,6.25,0,0)"><stop offset="0" style="stop-color:#000;stop-opacity:.101961"/><stop offset="1" style="stop-color:#000;stop-opacity:0"/></linearGradient><linearGradient id="linear1" gradientUnits="userSpaceOnUse" x1="4.004" y1="18.529" x2="4.047" y2="18.573" gradientTransform="matrix(6.25,0,0,6.25,0,0)"><stop offset="0" style="stop-color:#000;stop-opacity:.101961"/><stop offset="1" style="stop-color:#000;stop-opacity:0"/></linearGradient><linearGradient id="linear2" gradientUnits="userSpaceOnUse" x1=".617" y1="5.771" x2="23.69" y2="16.53" gradientTransform="matrix(6.25,0,0,6.25,0,0)"><stop offset="0" style="stop-color:#fff;stop-opacity:.2"/><stop offset="1" style="stop-color:#fff;stop-opacity:0"/></linearGradient></defs><g id="surface1"><path style="stroke:none;fill-rule:nonzero;fill:#303c42;fill-opacity:1" d="M138.570312 45.789062C139.492188 39.613281 138.867188 23.738281 133.167969 8.292969 132.660156 6.945312 131.300781 6.113281 129.867188 6.273438 129.269531 6.34375 115.1875 8.164062 94.039062 23.46875 89.445312 22.363281 82.632812 21.742188 75 21.742188s-14.4375.625-19.042969 1.726562C34.800781 8.15625 20.71875 6.34375 20.117188 6.273438 18.6875 6.117188 17.332031 6.949219 16.820312 8.292969 11.117188 23.738281 10.5 39.613281 11.425781 45.789062 3.742188 55.289062.0 66.230469.0 79.179688c0 41.632812 21.824219 64.558593 61.457031 64.558593L75 143.75 88.539062 143.738281C128.167969 143.738281 150 120.8125 150 79.179688c0-12.949219-3.742188-23.890626-11.429688-33.390626zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:#5c6671;fill-opacity:1" d="M88.539062 137.488281 75 137.5 61.457031 137.488281C36.273438 137.488281 6.25 127.367188 6.25 79.179688c0-12.109376 3.46875-21.816407 10.925781-30.535157.71875-.875.925781000000001-2.058593.550781000000001-3.125C17 43.4375 16.71875 28.382812 21.800781 12.988281c4.78125 1.207031 16.300781 5.050781 31.65625 16.398438C54.25 29.96875 55.269531 30.15625 56.207031 29.867188c3.9375-1.171876 10.960938-1.875 18.792969-1.875 7.835938.0 14.863281.703124000000003 18.789062 1.871093C94.730469 30.140625 95.746094 29.964844 96.539062 29.382812c15.347657-11.34375 26.867188-15.1875 31.648438-16.394531 5.082031 15.394531 4.804688 30.449219 4.082031 32.53125C131.867188 46.585938 132.078125 47.785156 132.820312 48.648438 140.273438 57.367188 143.75 67.070312 143.75 79.179688c0 48.1875-30.023438 58.308593-55.210938 58.308593zm0 0"/><use xlink:href="#surface5" mask="url(#mask0)"/><path style="stroke:none;fill-rule:nonzero;fill:#303c42;fill-opacity:1" d="M89.488281 63.144531C84.886719 63.507812 80.125 63.875 75.226562 63.875H74.78125C69.875 63.875 65.117188 63.5 60.519531 63.144531 47.898438 62.164062 35.988281 61.230469 27.894531 69.429688 21.824219 75.59375 18.75 83.28125 18.75 92.261719 18.75 128.238281 46.429688 131.25 75.226562 131.25c28.34375.0 56.023438-3.011719 56.023438-38.988281.0-8.988281-3.074219-16.667969-9.144531-22.835938-8.085938-8.195312-20.003907-7.269531-32.617188-6.28125zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:#d9cfcc;fill-opacity:1" d="M74.773438 125C42.492188 125 25 119.78125 25 92.261719c0-7.285157 2.476562-13.492188 7.34375-18.449219 3.945312-4 9.382812-5.074219 16.039062-5.074219C51.925781 68.738281 55.824219 69.042969 60.023438 69.367188c4.71875.375 9.589843.71875 14.84375.746093000000002C80.335938 70.230469 85.210938 69.738281 89.976562 69.367188c12.09375-.929687999999999 21.628907-1.6875 27.671876 4.4375C122.523438 78.769531 125 84.976562 125 92.261719 125 119.78125 107.507812 125 74.773438 125zm0 0"/><use xlink:href="#surface8" mask="url(#mask1)"/><path style="stroke:none;fill-rule:nonzero;fill:#303c42;fill-opacity:1" d="M50 75c-7.125.0-12.5 9.40625-12.5 21.875S42.875 118.75 50 118.75s12.5-9.40625 12.5-21.875S57.125 75 50 75zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:#6d4c41;fill-opacity:1" d="M50 112.5c-2.550781.0-6.25-6.085938-6.25-15.625S47.449219 81.25 50 81.25s6.25 6.085938 6.25 15.625S52.550781 112.5 50 112.5zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:#303c42;fill-opacity:1" d="M1e2 75c-7.125.0-12.5 9.40625-12.5 21.875S92.875 118.75 1e2 118.75s12.5-9.40625 12.5-21.875S107.125 75 1e2 75zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:#6d4c41;fill-opacity:1" d="M1e2 112.5c-2.550781.0-6.25-6.085938-6.25-15.625S97.449219 81.25 1e2 81.25s6.25 6.085938 6.25 15.625S102.550781 112.5 1e2 112.5zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:#a1887f;fill-opacity:1" d="M52.21875 90.511719c0 1.539062-1.246094 2.789062-2.789062 2.789062-1.539063.0-2.785157-1.25-2.785157-2.789062.0-1.539063 1.246094-2.785157 2.785157-2.785157 1.542968.0 2.789062 1.246094 2.789062 2.785157zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:#a1887f;fill-opacity:1" d="M102.21875 90.511719c0 1.539062-1.246094 2.789062-2.789062 2.789062-1.539063.0-2.785157-1.25-2.785157-2.789062.0-1.539063 1.246094-2.785157 2.785157-2.785157 1.542968.0 2.789062 1.246094 2.789062 2.785157zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:url(#linear0)" d="M125 72.8125 124.894531 72.855469C129.039062 78.382812 131.25 84.851562 131.25 92.261719c0 35.976562-27.679688 38.988281-56.023438 38.988281-16.664062.0-32.914062-1.054688-43.683593-8.875l13.386719 13.386719c5.507812 1.207031 11.121093 1.71875 16.527343 1.71875L75 137.5 88.539062 137.488281c22.992188.0 49.941407-8.53125 54.472657-46.664062zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:url(#linear1)" d="M25.070312 115.800781 25 115.832031 25.269531 116.101562zm0 0"/><path style="stroke:none;fill-rule:nonzero;fill:url(#linear2)" d="M138.570312 45.789062C139.492188 39.613281 138.867188 23.738281 133.167969 8.292969 132.660156 6.945312 131.300781 6.113281 129.867188 6.273438 129.269531 6.34375 115.1875 8.164062 94.039062 23.46875 89.445312 22.363281 82.632812 21.742188 75 21.742188s-14.4375.625-19.042969 1.726562C34.800781 8.15625 20.71875 6.34375 20.117188 6.273438 18.6875 6.117188 17.332031 6.949219 16.820312 8.292969 11.117188 23.738281 10.5 39.613281 11.425781 45.789062 3.742188 55.289062.0 66.230469.0 79.179688c0 41.632812 21.824219 64.558593 61.457031 64.558593L75 143.75 88.539062 143.738281C128.167969 143.738281 150 120.8125 150 79.179688c0-12.949219-3.742188-23.890626-11.429688-33.390626zm0 0"/></g></svg></a></li><li><a href=mailto:0x000216@gmail.com target=_blank title=Email rel=me><svg width="512" height="512" viewBox="0 0 512 512"><path d="M33.994 73.934C27.135 76.021 21.866 80.426 18.364 87 16.567 90.374 16.5 96.438 16.5 256V421.5l2.158 4C21.184 430.18 25.101 434.027 30 436.636 33.003 438.235 36.304 438.546 53.246 438.825L72.993 439.15l.253-137.978L73.5 163.194 90 175.45C99.075 182.191 140.17 212.762 181.321 243.386l74.821 55.68 21.179-15.752C288.97 274.65 330 244.021 368.5 215.248l70-52.312L438.754 301.043 439.007 439.15 458.754 438.825C476.726 438.529 478.859 438.306 482.5 436.342 487.18 433.816 491.027 429.899 493.636 425 495.433 421.626 495.5 415.562 495.5 256 495.5 96.438 495.433 90.374 493.636 87 488.565 77.48 482.099 73.799 469.397 73.202L460.293 72.774 454.397 77.337C439.883 88.566 256.423 221.935 255.772 221.73 254.976 221.479 78.972 93.309 62.177 80.75L51.812 73 44.156 73.086C39.945 73.133 35.372 73.515 33.994 73.934" stroke="none" fill="#d34c3c" fill-rule="evenodd"/><path d="M54.528 74.911C55.612 75.963 74.953 90.25 97.507 106.661c22.555 16.412 67.228 48.964 99.275 72.339s58.594 42.604 58.993 42.731C256.549 221.978 454.147 78.147 457.472 74.916 459.416 73.028 456.501 73 256 73 55.677 73 52.586 73.029 54.528 74.911M73 300.878V439H256 439V300.941C439 190.749 438.748 163.038 437.75 163.654 437.063 164.078 405.45 187.632 367.5 215.994c-37.95 28.363-78.528 58.655-90.173 67.316l-21.173 15.748-66.827-49.716C152.572 221.999 111.925 191.776 99 182.18s-24.062-17.892-24.75-18.436C73.251 162.954 73 190.519 73 300.878" stroke="none" fill="#e5e5e5" fill-rule="evenodd"/></svg></a></li><li><a href=https://nexus-security.github.io/index.xml target=_blank title=RSS rel=me><svg id="Layer_1" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 512 512" style="enable-background:new 0 0 512 512"><path style="fill:#f78c20" d="M78.333 355.334C35.14 355.334.0 390.474.0 433.667S35.14 512 78.333 512s78.333-35.14 78.333-78.333-35.14-78.333-78.333-78.333z"/><g><path style="fill:#ffa929" d="M78.333 381.445c-28.795.0-52.222 23.427-52.222 52.222s23.427 52.222 52.222 52.222 52.222-23.427 52.222-52.222-23.427-52.222-52.222-52.222z"/><path style="fill:#ffa929" d="M477.918 264.861c-21.843-51.641-53.111-98.019-92.936-137.842-39.824-39.824-86.201-71.093-137.843-92.935C193.669 11.468 136.874.0 78.333.0c-4.807.0-8.704 3.897-8.704 8.704v85.519c0 4.807 3.897 8.704 8.704 8.704 182.37.0 330.74 148.369 330.74 330.74.0 4.807 3.897 8.704 8.704 8.704h85.52c4.807.0 8.704-3.897 8.704-8.704C512 375.126 500.533 318.331 477.918 264.861z"/><path style="fill:#ffa929" d="M78.333 163.853c-4.807.0-8.704 3.897-8.704 8.704v95.74c0 4.807 3.897 8.704 8.704 8.704 86.386.0 156.666 70.281 156.666 156.666.0 4.807 3.897 8.704 8.704 8.704h95.74c4.807.0 8.704-3.897 8.704-8.704.0-72.07-28.065-139.826-79.027-190.787-50.961-50.961-118.717-79.027-190.787-79.027z"/></g><g><path style="fill:#f78c20" d="M78.333 242.186c-2.918.0-5.817.076-8.704.206v25.905c0 4.807 3.897 8.704 8.704 8.704 86.386.0 156.666 70.281 156.666 156.666.0 4.807 3.897 8.704 8.704 8.704h25.905c.129-2.886.206-5.786.206-8.704.0-105.752-85.729-191.481-191.481-191.481z"/><path style="fill:#f78c20" d="M78.333 68.113c-2.91.0-5.81.042-8.704.11v26.001c0 4.807 3.897 8.704 8.704 8.704 182.37.0 330.74 148.369 330.74 330.74.0 4.807 3.897 8.704 8.704 8.704h26.001c.067-2.894.11-5.793.11-8.704C443.887 231.777 280.223 68.113 78.333 68.113z"/></g><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/></svg></a></li><li><a href=https://x.com/0x000216 target=_blank title=Twitter rel=me><svg xmlns:xlink="http://www.w3.org/1999/xlink" width="150" height="150" viewBox="0 0 150 150"><g id="surface1"><path style="stroke:none;fill-rule:nonzero;fill:#03a9f4;fill-opacity:1" d="M150 33.179688c-5.640625 2.460937-11.609375 4.09375-17.71875 4.855468 6.4375-3.820312 11.25-9.867187 13.527344-16.996094-6.027344 3.574219-12.621094 6.089844-19.5 7.441407-8.628906-9.210938-22.007813-12.214844-33.746094-7.574219-11.738281 4.636719-19.449219 15.980469-19.445312 28.601562.0 2.4375.203124000000003 4.78125.710937000000001 7.015626C49.085938 55.308594 26.03125 43.609375 10.445312 24.355469 2.25 38.402344 6.386719 56.398438 19.894531 65.457031 15.027344 65.324219 10.261719 64.027344 6 61.667969V62.007812c.015625 14.636719 10.304688 27.25 24.636719 30.214844-2.628907.691406000000001-5.339844 1.03125-8.0625 1.011719C20.621094 93.269531 18.667969 93.09375 16.753906 92.710938c4.070313 12.507812 15.585938 21.089843 28.734375 21.421874-10.886719 8.511719-24.308593 13.128907-38.128906 13.113282C4.835938 127.246094 2.417969 127.132812.0 126.824219c14.0625 9.0625 30.445312 13.855469 47.175781 13.800781 56.585938.0 87.523438-46.875 87.523438-87.507812.0-1.359376-.046875-2.671876-.113281000000001-3.972657C140.652344 44.804688 145.875 39.394531 150 33.179688zm0 0"/></g></svg></a></li><li><a href=https://www.minds.com/0x000216/ target=_blank title=Minds rel=me><svg id="Capa_1" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 512 512" style="enable-background:new 0 0 512 512"><g><g><path style="fill:#ffe1b2" d="M256 33.085C245.078 13.38 224.079.0 2e2.0c-23.781.0-45.57 13.293-56.594 34.184C115.711 41.602 96 66.977 96 96c0 .059.0.113.0.172-9.977 7.512-16 19.301-16 31.828.0 1.316.078 2.637.234 3.992C60.211 145.266 48 167.758 48 192c0 14.07 4.039 27.543 11.719 39.262C57.273 236.512 56 242.207 56 248c0 2.738.281 5.445.828 8.098C36.672 267.308 24 288.539 24 312c0 27.973 18.305 52.34 44.109 60.785C65.398 378.828 64 385.324 64 392c0 21.098 13.805 39.508 33.539 45.727C103.891 466.746 129.828 488 160 488c4.617.0 9.227-.512 13.766-1.527C181.992 502 198.141 512 216 512c16.687.0 31.396-8.567 40-21.523V33.085z"/></g><g><g><path style="fill:#ffb980" d="M264 256c-4.422.0-8-3.582-8-8 0-22.055-17.945-40-40-40-8.008.0-15.734 2.355-22.336 6.812-3.023 2.043-7.055 1.781-9.797-.652-3.156-2.809-8.477-6.16-15.867-6.16-4.422.0-8-3.582-8-8s3.578-8 8-8c7.711.0 15.234 2.293 21.719 6.539C197.773 194.246 206.758 192 216 192c30.875.0 56 25.121 56 56C272 252.418 268.422 256 264 256z"/></g></g><g><g><path style="fill:#ffb980" d="M120 120c18.977.0 36.875 7.312 50.414 20.594 3.141 3.09 8.203 3.047 11.312-.109 3.094-3.152 3.047-8.219-.109-11.312C165.07 112.941 143.187 104 120 104c-13.046.0-25.395 2.93-36.542 8.046C81.253 117.019 80 122.423 80 128c0 1.316.078 2.637.234 3.992-.094.062-.173.139-.267.202C91.423 124.501 105.193 120 120 120z"/></g></g><g><g><path style="fill:#ffb980" d="M216 360c0-4.418-3.578-8-8-8s-8 3.582-8 8c0 17.645-14.352 32-32 32-14.211.0-26.82-9.648-30.664-23.465-.703-2.512-2.578-4.523-5.039-5.395-2.453-.871-5.188-.492-7.305 1.02C114.094 371.906 101.305 376 88 376c-6.948.0-13.625-1.149-19.894-3.207-2.214 4.939-3.501 10.19-3.916 15.586C71.714 390.73 79.711 392 88 392c13.297.0 26.187-3.266 37.773-9.52C133.969 397.894 150.141 408 168 408c26.469.0 48-21.531 48-48z"/></g></g><g><path style="fill:#fdc88e" d="M488 312c0-23.461-12.672-44.691-32.828-55.902.547-2.652.828-5.359.828-8.098.0-5.793-1.273-11.488-3.719-16.738C459.961 219.543 464 206.07 464 192c0-24.242-12.211-46.734-32.234-60.008.156-1.355.234-2.676.234-3.992.0-12.527-6.023-24.316-16-31.828.0-.059.0-.113.0-.172.0-29.023-19.711-54.398-47.406-61.816C357.57 13.293 335.781.0 312 0c-24.08.0-45.078 13.38-56 33.085v457.391C264.604 503.433 279.313 512 296 512c17.859.0 34.008-10 42.234-25.527C342.773 487.488 347.383 488 352 488c30.172.0 56.109-21.254 62.461-50.273C434.195 431.508 448 413.097 448 392c0-6.676-1.398-13.172-4.109-19.215C469.695 364.34 488 339.973 488 312z"/></g><g><path style="fill:#f8ab6b" d="M272.008 151.199C272 151.465 272 151.734 272 152c0 26.469 21.531 48 48 48s48-21.531 48-48c0-4.418-3.578-8-8-8s-8 3.582-8 8c0 17.645-14.352 32-32 32s-32-14.355-32-32c0-2.184.219-4.359.656-6.465.492-2.395-.133-4.883-1.703-6.754-1.57-1.871-4.016-3.066-6.352-2.859-.453.012-.891.059-.602.078-13.234.0-24-10.766-24-24v31.813C260.673 147.348 266.061 149.988 272.008 151.199z"/></g><g><path style="fill:#f8ab6b" d="M296 328c9.242.0 18.219-2.246 26.281-6.539C328.765 325.707 336.289 328 344 328c4.422.0 8-3.582 8-8s-3.578-8-8-8c-7.391.0-12.711-3.352-15.867-6.16-2.742-2.434-6.766-2.695-9.797-.656C311.726 309.644 304 312 296 312c-22.055.0-40-17.945-40-40v39.116C266.174 321.517 280.337 328 296 328z"/></g><g><g><path style="fill:#f8ab6b" d="M431.765 131.992c.156-1.355.234-2.676.234-3.992.0-5.577-1.253-10.981-3.458-15.954C417.395 106.93 405.046 104 392 104c-4.422.0-8 3.582-8 8s3.578 8 8 8c14.807.0 28.577 4.501 40.032 12.194C431.939 132.131 431.859 132.054 431.765 131.992z"/></g></g><g><g><path style="fill:#f8ab6b" d="M447.81 388.38c-.415-5.396-1.702-10.647-3.916-15.586C437.624 374.85 430.948 376 424 376c-13.578.0-26.594-4.266-37.641-12.332-2.07-1.5-4.719-1.93-7.133-1.168-2.43.77-4.344 2.648-5.164 5.059C369.101 382.176 355.414 392 340 392c-4.422.0-8 3.582-8 8s3.578 8 8 8c18.875.0 35.961-10.191 45.094-26.156C396.976 388.512 410.258 392 424 392 432.288 392 440.285 390.73 447.81 388.38z"/></g></g></g><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/></svg></a></li><li><a href=https://www.buymeacoffee.com/0x000216 target=_blank title=Coffee rel=me><svg id="Capa_1" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 340 340" style="enable-background:new 0 0 340 340"><g id="XMLID_18_"><polygon id="XMLID_138_" style="fill:#dedde0" points="76.429,290 80,340 170,340 170,290"/><polygon id="XMLID_169_" style="fill:#dedde0" points="170,80 61.429,80 65,130 170,130"/><polygon id="XMLID_197_" style="fill:#acabb1" points="170,290 170,340 260,340 263.571,290"/><polygon id="XMLID_221_" style="fill:#acabb1" points="170,80 170,130 275,130 278.571,80"/><path id="XMLID_222_" style="fill:#ffda44" d="M170 260c-22.091.0-40-22.386-40-50s17.909-50 40-50v-30H65 50l10 160h16.429H170V260z"/><path id="XMLID_33_" style="fill:#ff9811" d="M170 130v30c22.091.0 40 22.386 40 50s-17.909 50-40 50v30h93.571H280l10-160h-15H170z"/><path id="XMLID_223_" style="fill:#50412e" d="M210 210c0-27.614-17.909-50-40-50v1e2c22.091.0 40-22.386 40-50z"/><path id="XMLID_224_" style="fill:#786145" d="M130 210c0 27.614 17.909 50 40 50V160c-22.091.0-40 22.386-40 50z"/><polygon id="XMLID_225_" style="fill:#50412e" points="278.571,80 300,80 300,40 260,40 260,0 80,0 80,40 40,40 40,80 61.429,80 170,80"/></g><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/><g/></svg></a></li></ol><ol class=menu id=main-menu><li><a href=/><svg class="icon icon-tabler icon-tabler-home" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><polyline points="5 12 3 12 12 3 21 12 19 12"/><path d="M5 12v7a2 2 0 002 2h10a2 2 0 002-2v-7"/><path d="M9 21v-6a2 2 0 012-2h2a2 2 0 012 2v6"/></svg>
<span>Home</span></a></li><li><a href=/about/><svg class="icon icon-tabler icon-tabler-user" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="12" cy="7" r="4"/><path d="M6 21v-2a4 4 0 014-4h4a4 4 0 014 4v2"/></svg>
<span>About Us</span></a></li><li><a href=/termsofservice/><svg class="icon icon-tabler icon-tabler-pencil" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M4 20h4L18.5 9.5a1.5 1.5.0 00-4-4L4 16v4"/><line x1="13.5" y1="6.5" x2="17.5" y2="10.5"/></svg>
<span>Terms Of Service</span></a></li><li><a href=/privacypolicy/><svg class="icon icon-tabler icon-tabler-archive" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><rect x="3" y="4" width="18" height="4" rx="2"/><path d="M5 8v10a2 2 0 002 2h10a2 2 0 002-2V8"/><line x1="10" y1="12" x2="14" y2="12"/></svg>
<span>Privacy Policy</span></a></li><li><a href=/disclaimer/><svg class="icon icon-tabler icon-tabler-link" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><path d="M10 14a3.5 3.5.0 005 0l4-4a3.5 3.5.0 00-5-5l-.5.5"/><path d="M14 10a3.5 3.5.0 00-5 0l-4 4a3.5 3.5.0 005 5l.5-.5"/></svg>
<span>Disclaimer</span></a></li><li><a href=/search/><svg class="icon icon-tabler icon-tabler-search" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="10" cy="10" r="7"/><line x1="21" y1="21" x2="15" y2="15"/></svg>
<span>Search</span></a></li><li><a href=/contact/><svg class="icon icon-tabler icon-tabler-mail" width="44" height="44" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><rect x="3" y="5" width="18" height="14" rx="2"/><polyline points="3 7 12 13 21 7"/></svg>
<span>Contact Us</span></a></li><li class=menu-bottom-section><ol class=menu><li id=dark-mode-toggle><svg class="icon icon-tabler icon-tabler-toggle-left" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="8" cy="12" r="2"/><rect x="2" y="6" width="20" height="12" rx="6"/></svg>
<svg class="icon icon-tabler icon-tabler-toggle-right" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="16" cy="12" r="2"/><rect x="2" y="6" width="20" height="12" rx="6"/></svg>
<span>Dark Mode</span></li></ol></li></ol></aside><aside class="sidebar right-sidebar sticky"><section class="widget archives"><div class=widget-icon><svg class="icon icon-tabler icon-tabler-hash" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><line x1="5" y1="9" x2="19" y2="9"/><line x1="5" y1="15" x2="19" y2="15"/><line x1="11" y1="4" x2="7" y2="20"/><line x1="17" y1="4" x2="13" y2="20"/></svg></div><h2 class="widget-title section-title">Table of contents</h2><div class=widget--toc><nav id=TableOfContents><ol><li><a href=#output>Output</a></li><li><a href=#materials-and-methods>Materials and Methods</a><ol><li><a href=#the-pdb-format>The PDB Format</a><ol><li><a href=#title><code>TITLE</code></a></li><li><a href=#atom><code>ATOM</code></a></li><li><a href=#ter><code>TER</code></a></li><li><a href=#model-and-endmdl><code>MODEL</code> and <code>ENDMDL</code></a></li><li><a href=#ssbond><code>SSBOND</code></a></li><li><a href=#conect><code>CONECT</code></a></li><li><a href=#source><code>SOURCE</code></a></li></ol></li></ol></li><li><a href=#nr-format>NR Format</a></li><li><a href=#database-construction-and-parsing-data>Database Construction and Parsing Data</a><ol><li><a href=#downloading-data>Downloading Data</a></li><li><a href=#parsing-data>Parsing Data</a></li><li><a href=#database-construction>Database Construction</a><ol><li><a href=#tables>Tables</a></li></ol></li></ol></li><li><a href=#filtering-and-processing-data>Filtering and Processing Data</a></li><li><a href=#postgresql-as-a-processing-intermediary>PostgreSQL as a Processing Intermediary</a></li></ol></nav></div></section></aside><main class="main full-width"><article class=main-article><header class=article-header><div class=article-details><div class=article-title-wrapper><h2 class=article-title><a href=/creating-a-bioinformatics-database-for-studying-disulfide-bonds/>Creating a bioinformatics database for studying disulfide bonds</a></h2></div><footer class=article-time><div><svg class="icon icon-tabler icon-tabler-calendar-time" width="56" height="56" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><path d="M11.795 21H5a2 2 0 01-2-2V7a2 2 0 012-2h12a2 2 0 012 2v4"/><circle cx="18" cy="18" r="4"/><path d="M15 3v4"/><path d="M7 3v4"/><path d="M3 11h16"/><path d="M18 16.496V18l1 1"/></svg>
<time class=article-time--published>Feb 26, 2024</time></div><div><svg class="icon icon-tabler icon-tabler-clock" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="12" cy="12" r="9"/><polyline points="12 7 12 12 15 15"/></svg>
<time class=article-time--reading>26 minute read</time></div></footer></div></header><section class=article-content><p>The Protein Data Bank (PDB) serves as a central resource for experimentally-determined structures of proteins, nucleic acids, and complex assemblies. All the data housed in the PDB is obtained through experimental techniques such as X-ray crystallography, spectroscopy, NMR, and other methods.</p><p>This article outlines a methodology for extracting, filtering, and cleaning data from the PDB, ultimately facilitating the kind of analysis detailed in the article <a class=link href=https://doi.org/10.1093/protein/gzt063 target=_blank rel=noopener>Occurrence of protein disulfide bonds in different domains of life: a comparison of proteins from the Protein Data Bank</a>, featured in <em>Protein Engineering, Design and Selection</em>, Volume 27, Issue 3, 1 March 2014, pp. 65–72.</p><p>The PDB often includes numerous repeating structures with variations in resolution, methodologies, mutations, and other factors. Conducting experiments with identical or highly similar proteins can introduce bias into group analysis. To address this, we need to carefully select an appropriate structure from a set of duplicates, necessitating the use of a non-redundant (NR) set of proteins.</p><p>For normalization purposes, I suggest downloading the <a class=link href=ftp://ftp.wwpdb.org/pub/pdb/data/monomers/het_dictionary.txt>chemical compound dictionary</a> for importing atom names into a database that uses 3NF or uses a star schema and dimensional modeling. (I’ve also used <a class=link href=https://swift.cmbi.umcn.nl/gv/dssp/DSSP_3.html target=_blank rel=noopener>DSSP</a> to aid in the removal of problematic structures. While I won&rsquo;t delve into the details in this article, it&rsquo;s worth noting that I haven&rsquo;t utilized any other DSSP features.)</p><p>The data employed in this study comprises single-unit proteins from various species that contain at least one disulfide bond. For analysis, all disulfide bonds are categorized as consecutive or nonconsecutive, categorized by domain (archaea, prokaryote, viral, eukaryote, or other), and classified by length.</p><div data-testid=image-container class="X3Bg-O9c _1xIzGpI7"><figure class=_1sK7TguO><picture class=!contents data-testid=picture><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=360, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=360&amp;dpr=2 2x" media="(max-width: 360px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=480, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=480&amp;dpr=2 2x" media="(min-width: 360.1px) and (max-width: 480px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=768, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=768&amp;dpr=2 2x" media="(min-width: 480.1px) and (max-width: 768px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=1024, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=1024&amp;dpr=2 2x" media="(min-width: 768.1px) and (max-width: 1024px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=1440, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=1440&amp;dpr=2 2x" media="(min-width: 1024.1px) and (max-width: 1440px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=1920, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png&amp;width=1920&amp;dpr=2 2x" media="(min-width: 1440.1px) and (max-width: 1920px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png" media="(min-width: 1920.1px)"><img alt="Primary and tertiary protein structures" src="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126730%2Ftoptal-blog-image-1533119338949-4e2712429ab9345b9a69919e84c483dc.png" loading=lazy decoding=async class=Djd3YO_2></picture></figure></div><div class="pop_out_box is-full_width" style=font-size:1em;text-align:center>Primary and tertiary protein structures, before and after protein folding.<br>Source: <i>Protein Engineering, Design and Selection</i>, as mentioned at the beginning of this article.</div><h2 id=output>Output</h2><p>To facilitate analysis in tools like R or SPSS, the data needs to be structured in a database table as follows:</p><div class=responsive-table><div class=table-wrapper><table><thead><tr><th>Column</th><th>Type</th><th>Description</th></tr></thead><tbody><tr><td><code>code</code></td><td><code>character(4)</code></td><td>Experiment ID (alphanumeric, case-insensitive, and cannot start with a zero)</td></tr><tr><td><code>title</code></td><td><code>character varying(1000)</code></td><td>Title of the experiment, for reference (field can be also text format)</td></tr><tr><td><code>ss_bonds</code></td><td><code>integer</code></td><td>Number of disulfide bonds in the chosen chain</td></tr><tr><td><code>ssbonds_overlap</code></td><td><code>integer</code></td><td>Number of overlapping disulfide bonds</td></tr><tr><td><code>intra_count</code></td><td><code>integer</code></td><td>Number of bonds made within the same chain</td></tr><tr><td><code>sci_name_src</code></td><td><code>character varying(5000)</code></td><td>Scientific name of organism from which the sequence is taken</td></tr><tr><td><code>tax_path</code></td><td><code>character varying</code></td><td>Path in Linnaean classification tree</td></tr><tr><td><code>src_class</code></td><td><code>character varying(20)</code></td><td>Top-level class of organism (eukaryote, prokaryote, virus, archaea, other)</td></tr><tr><td><code>has_reactives7</code></td><td><code>boolean</code></td><td>True if and only if the sequence contains reactive centers</td></tr><tr><td><code>len_class7</code></td><td><code>integer</code></td><td>Length of sequence in set 7 (set with p-value 10e-7 calculated by blast)</td></tr></tbody></table></div></div><h2 id=materials-and-methods>Materials and Methods</h2><p>To achieve this objective, the first step involves gathering data from <a class=link href=https://www.rcsb.org target=_blank rel=noopener>rcsb.org</a>, a repository offering downloadable PDB structures of experiments in various formats.</p><p>While data is stored in multiple formats, this example will focus on the formatted fixed-space delimited textual format (PDB). An alternative to the PDB textual format is PDBML, its XML counterpart, but it occasionally exhibits issues with malformed atom position naming entries, potentially hindering data analysis. Although older formats like mmCIF might be accessible, they won&rsquo;t be covered in this article.</p><h3 id=the-pdb-format>The PDB Format</h3><p>The PDB format utilizes a fragmented fixed-width textual format, easily parsed by SQL queries, Java plugins, or Perl modules, for instance. Each data type within the file container is represented by a line commencing with the corresponding tag—we will explore each tag type in the subsequent subsections. The line length is capped at 80 characters, with tags occupying a maximum of six characters plus one or more spaces, collectively consuming eight bytes. However, there are instances where spaces are omitted between tags and data, typically for <code>CONECT</code> tags.</p><h4 id=title><code>TITLE</code></h4><p>The <code>TITLE</code> tag designates a line as part of the experiment&rsquo;s title, encompassing the molecule name and other pertinent information like the insertion, mutation, or deletion of a specific amino acid.</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span><span class=lnt>4
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-fallback data-lang=fallback><span class=line><span class=cl>12345678901234567890123456789012345678901234567890123456789012345678901234567890
</span></span><span class=line><span class=cl>TITLE     A TWO DISULFIDE DERIVATIVE OF CHARYBDOTOXIN WITH DISULFIDE            
</span></span><span class=line><span class=cl>TITLE    2 13-33 REPLACED BY TWO ALPHA-AMINOBUTYRIC ACIDS, NMR, 30              
</span></span><span class=line><span class=cl>TITLE    3 STRUCTURES                                                           
</span></span></code></pre></td></tr></table></div></div><p>When a <code>TITLE</code> record spans multiple lines, the title needs to be concatenated based on a continuation number, right-aligned on bytes 9 and 10 (depending on the number of lines).</p><h4 id=atom><code>ATOM</code></h4><p><code>ATOM</code> lines contain coordinate data for each atom in an experiment. Due to insertions, mutations, alternate locations, or multiple models within an experiment, the same atom might be repeated multiple times. The process of selecting the correct atoms will be explained later.</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span><span class=lnt>4
</span><span class=lnt>5
</span><span class=lnt>6
</span><span class=lnt>7
</span><span class=lnt>8
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-fallback data-lang=fallback><span class=line><span class=cl>12345678901234567890123456789012345678901234567890123456789012345678901234567890
</span></span><span class=line><span class=cl>ATOM    390  N   GLY A  26      -1.120  -2.842   4.624  1.00  0.00           N  
</span></span><span class=line><span class=cl>ATOM    391  CA  GLY A  26      -0.334  -2.509   3.469  1.00  0.00           C  
</span></span><span class=line><span class=cl>ATOM    392  C   GLY A  26       0.682  -1.548   3.972  1.00  0.00           C  
</span></span><span class=line><span class=cl>ATOM    393  O   GLY A  26       0.420  -0.786   4.898  1.00  0.00           O  
</span></span><span class=line><span class=cl>ATOM    394  H   GLY A  26      -0.832  -2.438   5.489  1.00  0.00           H  
</span></span><span class=line><span class=cl>ATOM    395  HA2 GLY A  26       0.163  -3.399   3.111  1.00  0.00           H  
</span></span><span class=line><span class=cl>ATOM    396  HA3 GLY A  26      -0.955  -2.006   2.739  1.00  0.00           H  
</span></span></code></pre></td></tr></table></div></div><p>The above example originates from experiment <code>1BAH</code>. The initial column indicates the record type, while the second column represents the atom&rsquo;s serial number, unique for each atom in a structure.</p><p>Next to the serial number lies the four-byte atom position label, from which the element&rsquo;s chemical symbol can be extracted, although it might not always be present as a separate column.</p><p>Following the atom name is a three-letter residue code, corresponding to an amino acid in the case of proteins. Subsequently, the chain is represented by a single letter. A <em>chain</em> refers to a single sequence of amino acids, with or without gaps. Ligands can also be assigned to a chain, detectable by significant gaps in the amino acid sequence, detailed in the next column. When a structure includes mutations, the insertion code, a letter distinguishing the affected residue, is present in an additional column after the sequence column.</p><p>The following three columns contain the spatial coordinates of each atom, expressed in Angstroms (Å). Adjacent to these coordinates is the occupancy column, indicating the probability (on a scale of zero to one) of finding the atom at that specific location.</p><p>The penultimate column, the temperature factor (measured in Å²), provides insights into the disorder within the crystal. <a class=link href=http://www.proxychem.com/macromolecular_crystallography.html target=_blank rel=noopener>A value greater than 60Å² signifies disorder, while one lower than 30Å² signifies confidence</a>. Its presence in PDB files is not guaranteed, as it depends on the experimental method employed.</p><p>The final columns—symbol and charge—are frequently absent. As mentioned earlier, the chemical symbol can be derived from the atom position column. When present, the charge is appended to the symbol as an integer followed by <code>+</code> or <code>-</code>, for example, <code>N1+</code>.</p><h4 id=ter><code>TER</code></h4><p>This tag denotes the end of a chain. Although the <code>TER</code> line is often omitted, as chain endings can be readily identified even without it.</p><h4 id=model-and-endmdl><code>MODEL</code> and <code>ENDMDL</code></h4><p>A <code>MODEL</code> line, containing the model&rsquo;s serial number, indicates the start of a structure&rsquo;s model.</p><p>An <code>ENDMDL</code> line marks the end of all atomic lines within that model.</p><h4 id=ssbond><code>SSBOND</code></h4><p>These lines represent disulfide bonds between cysteine amino acids. Although disulfide bonds can occur in other residue types, this article focuses solely on amino acids, thus only considering cysteine. The following example is taken from experiment <code>132L</code>:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span><span class=lnt>4
</span><span class=lnt>5
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-fallback data-lang=fallback><span class=line><span class=cl>12345678901234567890123456789012345678901234567890123456789012345678901234567890
</span></span><span class=line><span class=cl>SSBOND   1 CYS A    6    CYS A  127                          1555   1555  2.01
</span></span><span class=line><span class=cl>SSBOND   2 CYS A   30    CYS A  115                          1555   1555  2.05
</span></span><span class=line><span class=cl>SSBOND   3 CYS A   64    CYS A   80                          1555   1555  2.02
</span></span><span class=line><span class=cl>SSBOND   4 CYS A   76    CYS A   94                          1555   1555  2.02
</span></span></code></pre></td></tr></table></div></div><p>This example illustrates four disulfide bonds tagged in the file with their corresponding sequence numbers in the second column. All bonds involve cysteine (columns 3 and 6) and are located within chain <code>A</code> (columns 4 and 7). Each chain is followed by a residue sequence number indicating the bond&rsquo;s position within the peptide chain. While insertion codes are positioned next to each residue sequence, they are absent in this example because no amino acids were inserted in that region. The two penultimate columns are reserved for <a class=link href=https://web.archive.org/web/20150826124238/http://www.bmsc.washington.edu/CrystaLinks/man/pdb/part_74.html target=_blank rel=noopener>symmetry operations</a>, and the final column represents the distance between sulfur atoms, measured in Å.</p><p>Let&rsquo;s provide some context for this data.</p><p>The images below, generated using the rcsb.org NGL viewer, depict the structure of experiment <code>132L</code>, specifically showcasing a protein devoid of ligands. The first image utilizes a stick representation with CPK coloring, highlighting sulfurs and their bonds in yellow. V-shaped sulfur connections symbolize methionine connections, while Z-shaped connections denote disulfide bonds between cysteines.</p><div data-testid=image-container class="X3Bg-O9c _1xIzGpI7"><figure class=_1sK7TguO><picture class=!contents data-testid=picture><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=360, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=360&amp;dpr=2 2x" media="(max-width: 360px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=480, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=480&amp;dpr=2 2x" media="(min-width: 360.1px) and (max-width: 480px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=768, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=768&amp;dpr=2 2x" media="(min-width: 480.1px) and (max-width: 768px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=1024, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=1024&amp;dpr=2 2x" media="(min-width: 768.1px) and (max-width: 1024px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=1440, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=1440&amp;dpr=2 2x" media="(min-width: 1024.1px) and (max-width: 1440px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=1920, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png&amp;width=1920&amp;dpr=2 2x" media="(min-width: 1440.1px) and (max-width: 1920px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png" media="(min-width: 1920.1px)"><img alt="Stick representation with CPK coloring showing sulfur disulfide bonds in yellow" src="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126572%2Ftoptal-blog-image-1531208250548-476fde636a6da86f9f843bcf51d7119e.png" loading=lazy decoding=async class=Djd3YO_2></picture></figure></div><p>The subsequent image employs a simplified protein visualization technique called <em>backbone visualization</em>, colored by amino acids, with cysteines depicted in yellow. It portrays the same protein with excluded sidechains and only a portion of its peptide group included—in this instance, the protein backbone. Composed of three atoms: N-terminal, C-alpha, and C-terminal, this representation, while lacking disulfide bonds, provides a simplified view of the protein&rsquo;s spatial arrangement:</p><div data-testid=image-container class="X3Bg-O9c _1xIzGpI7"><figure class=_1sK7TguO><picture class=!contents data-testid=picture><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=360, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=360&amp;dpr=2 2x" media="(max-width: 360px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=480, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=480&amp;dpr=2 2x" media="(min-width: 360.1px) and (max-width: 480px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=768, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=768&amp;dpr=2 2x" media="(min-width: 480.1px) and (max-width: 768px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=1024, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=1024&amp;dpr=2 2x" media="(min-width: 768.1px) and (max-width: 1024px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=1440, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=1440&amp;dpr=2 2x" media="(min-width: 1024.1px) and (max-width: 1440px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=1920, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png&amp;width=1920&amp;dpr=2 2x" media="(min-width: 1440.1px) and (max-width: 1920px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png" media="(min-width: 1920.1px)"><img alt="Simplified protein backbone colored by amino acids where cysteines are yellow" src="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126573%2Ftoptal-blog-image-1531208278776-a8c6c618fff7e0ea145765e26b6fffce.png" loading=lazy decoding=async class=Djd3YO_2></picture></figure></div><p>Pipes are generated by connecting peptide-bonded atoms with a C-alpha atom. The color assigned to cysteine matches that of sulfur in the CPK coloring scheme. When cysteines are in close proximity, their sulfurs form disulfide bonds, enhancing the structure&rsquo;s stability. Without these bonds, the protein might exhibit excessive binding, leading to reduced stability at elevated temperatures.</p><h4 id=conect><code>CONECT</code></h4><p><code>CONECT</code> records denote connections between atoms. While these tags might be entirely absent in some cases, they can be present with complete data in others. Although potentially beneficial for disulfide bond analysis, they are not essential for this project, as non-tagged bonds are added by calculating distances. Including them would introduce unnecessary overhead and require verification.</p><h4 id=source><code>SOURCE</code></h4><p>These records provide information about the organism from which the molecule was extracted, incorporating subrecords for easier taxonomic classification. Similar to title records, they can span multiple lines.</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt> 1
</span><span class=lnt> 2
</span><span class=lnt> 3
</span><span class=lnt> 4
</span><span class=lnt> 5
</span><span class=lnt> 6
</span><span class=lnt> 7
</span><span class=lnt> 8
</span><span class=lnt> 9
</span><span class=lnt>10
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-fallback data-lang=fallback><span class=line><span class=cl>SOURCE    MOL_ID: 1;                                  
</span></span><span class=line><span class=cl>SOURCE   2 ORGANISM_SCIENTIFIC: ANOPHELES GAMBIAE;    
</span></span><span class=line><span class=cl>SOURCE   3 ORGANISM_COMMON: AFRICAN MALARIA MOSQUITO; 
</span></span><span class=line><span class=cl>SOURCE   4 ORGANISM_TAXID: 7165;                      
</span></span><span class=line><span class=cl>SOURCE   5 GENE: GST1-6;                              
</span></span><span class=line><span class=cl>SOURCE   6 EXPRESSION_SYSTEM: ESCHERICHIA COLI;       
</span></span><span class=line><span class=cl>SOURCE   7 EXPRESSION_SYSTEM_TAXID: 562;              
</span></span><span class=line><span class=cl>SOURCE   8 EXPRESSION_SYSTEM_STRAIN: BL21(DE3)PLYSS;  
</span></span><span class=line><span class=cl>SOURCE   9 EXPRESSION_SYSTEM_VECTOR_TYPE: PLASMID;    
</span></span><span class=line><span class=cl>SOURCE  10 EXPRESSION_SYSTEM_PLASMID: PXAGGST1-6      
</span></span></code></pre></td></tr></table></div></div><h2 id=nr-format>NR Format</h2><p>This section delves into the non-redundant (NR) chain PDB sets, with snapshots accessible at <a class=link href=ftp://ftp.ncbi.nih.gov/mmdb/nrtable/>ftp.ncbi.nih.gov/mmdb/nrtable/</a>. These sets aim to mitigate biases arising from protein similarity. NR comprises three sets with varying identity p-value levels generated by comparing all PDB structures. The results are compiled into textual files, which will be elaborated on subsequently. As not all columns are relevant to this project, only the essential ones will be discussed.</p><p>The first two columns contain the unique PDB experiment code and the chain identifier, as explained for <code>ATOM</code> records. Columns 6, 9, and C hold information regarding p-value representativeness, indicating the level of sequence similarity calculated using <a class=link href=http://guides.lib.berkeley.edu/ncbi/blast target=_blank rel=noopener>BLAST</a>. A value of zero signifies exclusion from a set, while a value of one indicates inclusion. These columns represent acceptance for sets with p-value cutoffs of 10e-7, 10e-40, and 10e-80, respectively. Only sets with a p-value cutoff of 10e-7 will be employed for analysis.</p><p>The final column indicates a structure&rsquo;s acceptability, where <code>a</code> signifies acceptable and <code>n</code> represents unacceptable.</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span><span class=lnt>4
</span><span class=lnt>5
</span><span class=lnt>6
</span><span class=lnt>7
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-fallback data-lang=fallback><span class=line><span class=cl>#---------------------------------------------------------------------------------------------------------------------------
</span></span><span class=line><span class=cl># 1  2       3     4    5 6     7    8 9     A    B C     D    E F    G      H      I      J     K     L   M   N     O P Q
</span></span><span class=line><span class=cl>#---------------------------------------------------------------------------------------------------------------------------
</span></span><span class=line><span class=cl>3F8V A   69715     1    1 1     1    1 1     1    1 1  9427    1 1   0.00   0.00   0.00   0.00  1.08   1   6   5   164 X a
</span></span><span class=line><span class=cl>3DKE X   68132     1    2 0     1    2 0     1    2 0 39139    1 1   0.00   0.00   0.00   0.00  1.25   1  11   7   164 X a
</span></span><span class=line><span class=cl>3HH3 A   77317     1    3 0     1    3 0     1    3 0    90    1 0   0.00   0.00   0.00   0.00  1.25   1   5   4   164 X a
</span></span><span class=line><span class=cl>3HH5 A   77319     1    4 0     1    4 0     1    4 0    90    2 0   0.00   0.00   0.00   0.00  1.25   1   4   4   164 X a
</span></span></code></pre></td></tr></table></div></div><h2 id=database-construction-and-parsing-data>Database Construction and Parsing Data</h2><p>Having established a basic understanding of the data and objectives, let&rsquo;s proceed with the practical aspects.</p><h3 id=downloading-data>Downloading Data</h3><p>All data required for this analysis can be found at the following three locations:</p><ul><li><a class=link href=ftp://ftp.wwpdb.org/pub/pdb/data/structures/divided/pdb/>ftp.wwpdb.org/pub/pdb/data/structures/divided/pdb/</a></li><li><a class=link href=ftp://ftp.ncbi.nih.gov/mmdb/nrtable/>ftp.ncbi.nih.gov/mmdb/nrtable/</a></li><li><a class=link href=ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdmp.zip>ftp.ncbi.nih.gov/pub/taxonomy/taxdmp.zip</a></li></ul><p>The first two links provide a list of archives. Using the most recent archive from each source is recommended to avoid issues stemming from insufficient resolution or quality. The third link directly provides the newest taxonomy archive.</p><h3 id=parsing-data>Parsing Data</h3><p>Parsing PDB files is typically accomplished using plugins or modules in languages like Java, Perl, or Python. In this study, I developed a custom Perl application without relying on pre-built PDB-parsing modules. This decision stems from my experience with parsing large datasets, where the most prevalent issue encountered with experimental data involves data errors. These errors can manifest as coordinate discrepancies, distance inconsistencies, incorrect line lengths, misplaced comments, and more.</p><p>The most effective approach to handling such situations is to initially store all data in the database as raw text. Common parsers are designed to handle ideal, specification-compliant data. However, in practice, data often deviates from the ideal, as will be illustrated in the filtering section, where the Perl import script is presented.</p><h3 id=database-construction>Database Construction</h3><p>When constructing the database, it&rsquo;s crucial to recognize that its primary purpose is data processing, with subsequent analysis conducted in SPSS or R. For our needs, PostgreSQL version 8.4 or higher is recommended.</p><p>The table structure closely mirrors that of the downloaded files, with minimal modifications. In this instance, the relatively small number of records makes normalization efforts unnecessary. As mentioned earlier, this database is intended for single-use purposes, meaning the tables are not <a class=link href=http://what-when-how.com/bioinformatics/relational-databases-in-bioinformatics/ target=_blank rel=noopener>built to be served on a website</a>. They serve as temporary structures for data processing. Once processing is complete, they can be discarded or retained as supplementary data, potentially for replication by other researchers.</p><p>In this specific scenario, the final output will be a single table, readily exportable to a file for analysis using statistical tools like SPSS or R.</p><h4 id=tables>Tables</h4><p>Data extracted from <code>ATOM</code> records must be linked to <code>HEADER</code> or <code>TITLE</code> records. The hierarchical relationship between data elements is illustrated in the diagram below.</p><div data-testid=image-container class="X3Bg-O9c _1xIzGpI7"><figure class=_1sK7TguO><picture class=!contents data-testid=picture><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=360, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=360&amp;dpr=2 2x" media="(max-width: 360px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=480, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=480&amp;dpr=2 2x" media="(min-width: 360.1px) and (max-width: 480px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=768, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=768&amp;dpr=2 2x" media="(min-width: 480.1px) and (max-width: 768px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=1024, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=1024&amp;dpr=2 2x" media="(min-width: 768.1px) and (max-width: 1024px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=1440, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=1440&amp;dpr=2 2x" media="(min-width: 1024.1px) and (max-width: 1440px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=1920, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png&amp;width=1920&amp;dpr=2 2x" media="(min-width: 1440.1px) and (max-width: 1920px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png" media="(min-width: 1920.1px)"><img alt="The natural hierarchy of disulfide bond data that would result in a database in third normal form" src="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126574%2Ftoptal-blog-image-1531234351923-9df0e52a15013a1dcfee2cad2ddd0c84.png" loading=lazy decoding=async class=Djd3YO_2></picture></figure></div><p>This diagram provides a simplified representation of a database in the third normal form (3NF), which introduces unnecessary overhead for our purposes. Calculating distances between atoms for disulfide bond detection would necessitate joins. In this case, we would have a table joined to itself twice, along with joins to secondary and primary structure tables, resulting in a time-consuming process. Given that secondary structure information is not always required for analysis, an alternative schema is proposed, particularly when data reuse or analysis of larger disulfide bond datasets is anticipated:</p><div data-testid=image-container class="X3Bg-O9c _1xIzGpI7"><figure class=_1sK7TguO><picture class=!contents data-testid=picture><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=360, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=360&amp;dpr=2 2x" media="(max-width: 360px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=480, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=480&amp;dpr=2 2x" media="(min-width: 360.1px) and (max-width: 480px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=768, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=768&amp;dpr=2 2x" media="(min-width: 480.1px) and (max-width: 768px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=1024, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=1024&amp;dpr=2 2x" media="(min-width: 768.1px) and (max-width: 1024px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=1440, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=1440&amp;dpr=2 2x" media="(min-width: 1024.1px) and (max-width: 1440px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=1920, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png&amp;width=1920&amp;dpr=2 2x" media="(min-width: 1440.1px) and (max-width: 1920px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png" media="(min-width: 1920.1px)"><img alt="A more detailed model of a database schema that does not use secondary structures (3NF)" src="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126731%2Ftoptal-blog-image-1533119410925-04dbc5ec57926b87c49cf6a1a3cb285f.png" loading=lazy decoding=async class=Djd3YO_2></picture></figure></div><p>While a warehouse model is not essential for disulfide bonds, which are less frequent than other covalent bonds, it could be employed. However, developing a star schema and dimensional modeling would be overly time-consuming and introduce complexities into queries:</p><div data-testid=image-container class="X3Bg-O9c _1xIzGpI7"><figure class=_1sK7TguO><picture class=!contents data-testid=picture><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=360, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=360&amp;dpr=2 2x" media="(max-width: 360px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=480, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=480&amp;dpr=2 2x" media="(min-width: 360.1px) and (max-width: 480px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=768, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=768&amp;dpr=2 2x" media="(min-width: 480.1px) and (max-width: 768px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=1024, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=1024&amp;dpr=2 2x" media="(min-width: 768.1px) and (max-width: 1024px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=1440, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=1440&amp;dpr=2 2x" media="(min-width: 1024.1px) and (max-width: 1440px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=1920, https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png&amp;width=1920&amp;dpr=2 2x" media="(min-width: 1440.1px) and (max-width: 1920px)"><source srcset="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png" media="(min-width: 1920.1px)"><img alt="The database layout using the star schema and dimensional modeling" src="https://assets.toptal.io/images?url=https%3A%2F%2Fuploads.toptal.io%2Fblog%2Fimage%2F126732%2Ftoptal-blog-image-1533119431254-0009dd8d1db64dadc39763a15d284c96.png" loading=lazy decoding=async class=Djd3YO_2></picture></figure></div><p>When processing all bonds is mandatory, I recommend adopting the star schema.</p><p>(Otherwise, it&rsquo;s unnecessary since disulfide bonds are less prevalent compared to other bonds. In this study, with approximately 30,000 disulfide bonds, processing in 3NF might suffice, but I opted for a non-normalized table approach, hence its exclusion from the diagram.)</p><p>Considering that the estimated number of all covalent bonds is at least twice the number of atoms in the tertiary structure, a 3NF approach would be highly inefficient. Consequently, denormalization using the star schema becomes necessary. In this schema, certain tables have two foreign key checks because bonds are formed between two atoms, requiring each atom to possess its own <code>primary_structure_id</code>, <code>atom_name_id</code>, and <code>residue_id</code>.</p><p>There are two methods for populating the <code>d_atom_name</code> dimension table: using data directly or utilizing the chemical component dictionary mentioned earlier. The dictionary follows a format similar to the PDB format, with <code>RESIDUE</code> and <code>CONECT</code> lines being the most relevant. The first column of <code>RESIDUE</code> contains a three-letter residue code, while <code>CONECT</code> includes the atom name and its connections, also represented by atom names. By parsing this file, we can extract all atom names and incorporate them into our database. However, it&rsquo;s advisable to accommodate the possibility of the database containing unlisted atom names.</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt> 1
</span><span class=lnt> 2
</span><span class=lnt> 3
</span><span class=lnt> 4
</span><span class=lnt> 5
</span><span class=lnt> 6
</span><span class=lnt> 7
</span><span class=lnt> 8
</span><span class=lnt> 9
</span><span class=lnt>10
</span><span class=lnt>11
</span><span class=lnt>12
</span><span class=lnt>13
</span><span class=lnt>14
</span><span class=lnt>15
</span><span class=lnt>16
</span><span class=lnt>17
</span><span class=lnt>18
</span><span class=lnt>19
</span><span class=lnt>20
</span><span class=lnt>21
</span><span class=lnt>22
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-fallback data-lang=fallback><span class=line><span class=cl>RESIDUE   PRO     17
</span></span><span class=line><span class=cl>CONECT      N      3 CA   CD   H   
</span></span><span class=line><span class=cl>CONECT      CA     4 N    C    CB   HA  
</span></span><span class=line><span class=cl>CONECT      C      3 CA   O    OXT 
</span></span><span class=line><span class=cl>CONECT      O      1 C   
</span></span><span class=line><span class=cl>CONECT      CB     4 CA   CG   HB2  HB3 
</span></span><span class=line><span class=cl>CONECT      CG     4 CB   CD   HG2  HG3 
</span></span><span class=line><span class=cl>CONECT      CD     4 N    CG   HD2  HD3 
</span></span><span class=line><span class=cl>CONECT      OXT    2 C    HXT 
</span></span><span class=line><span class=cl>CONECT      H      1 N   
</span></span><span class=line><span class=cl>CONECT      HA     1 CA  
</span></span><span class=line><span class=cl>CONECT      HB2    1 CB  
</span></span><span class=line><span class=cl>CONECT      HB3    1 CB  
</span></span><span class=line><span class=cl>CONECT      HG2    1 CG  
</span></span><span class=line><span class=cl>CONECT      HG3    1 CG  
</span></span><span class=line><span class=cl>CONECT      HD2    1 CD  
</span></span><span class=line><span class=cl>CONECT      HD3    1 CD  
</span></span><span class=line><span class=cl>CONECT      HXT    1 OXT 
</span></span><span class=line><span class=cl>END   
</span></span><span class=line><span class=cl>HET    PRO             17
</span></span><span class=line><span class=cl>HETNAM     PRO PROLINE
</span></span><span class=line><span class=cl>FORMUL      PRO    C5 H9 N1 O2
</span></span></code></pre></td></tr></table></div></div><p>In this project, coding speed takes precedence over execution speed and storage consumption. Therefore, I decided against normalization, as the primary goal is to generate a table with the columns outlined in the introduction.</p><p>This section will focus on explaining the most crucial tables.</p><p>The main tables include:</p><ul><li><code>proteins</code>: Stores experiment names and codes.</li><li><code>ps</code>: Contains the primary structure table, including <code>sequence</code>, <code>chain_id</code>, and <code>code</code>.</li><li><code>ts</code>: Holds the tertiary/quaternary structure extracted from raw data and transformed into the <code>ATOM</code> record format. It serves as a staging table and can be discarded after extraction. Ligands are excluded.</li><li><code>sources</code>: Lists the organisms from which experimental data was obtained.</li><li><code>tax_names</code>, <code>taxonomy_path</code>, <code>taxonomy_paths</code>: Contains Linnean taxonomy names from the NCBI taxonomy database, used to derive taxonomy paths for organisms listed in <code>sources</code>.</li><li><code>nr</code>: Stores the list of NCBI non-redundant proteins extracted from the NR set.</li><li><code>pdb_ssbond</code>: Contains the list of disulfide bonds within a given PDB file.</li></ul><h2 id=filtering-and-processing-data>Filtering and Processing Data</h2><p>Data retrieval is performed using snapshots from the RCSB PDB repository.</p><p>Each file is imported into a single table named <code>raw_pdb</code> in our PostgreSQL database using a Perl script, employing transactions of 10,000 inserts per chunk.</p><p>The structure of <code>raw_pdb</code> is as follows:</p><div class=responsive-table><div class=table-wrapper><table><thead><tr><th>Column</th><th>Type</th><th>Modifiers</th></tr></thead><tbody><tr><td>code</td><td>character varying(20)</td><td>not null</td></tr><tr><td>line_num</td><td>integer</td><td>not null</td></tr><tr><td>line_cont</td><td>character varying(80)</td><td>not null</td></tr></tbody></table></div></div><p>Below is the import script:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt> 1
</span><span class=lnt> 2
</span><span class=lnt> 3
</span><span class=lnt> 4
</span><span class=lnt> 5
</span><span class=lnt> 6
</span><span class=lnt> 7
</span><span class=lnt> 8
</span><span class=lnt> 9
</span><span class=lnt>10
</span><span class=lnt>11
</span><span class=lnt>12
</span><span class=lnt>13
</span><span class=lnt>14
</span><span class=lnt>15
</span><span class=lnt>16
</span><span class=lnt>17
</span><span class=lnt>18
</span><span class=lnt>19
</span><span class=lnt>20
</span><span class=lnt>21
</span><span class=lnt>22
</span><span class=lnt>23
</span><span class=lnt>24
</span><span class=lnt>25
</span><span class=lnt>26
</span><span class=lnt>27
</span><span class=lnt>28
</span><span class=lnt>29
</span><span class=lnt>30
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-perl data-lang=perl><span class=line><span class=cl><span class=ch>#!/usr/bin/perl -w</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=k>use</span> <span class=nn>FindBin</span> <span class=s>&#39;$Bin&#39;</span><span class=p>;</span>
</span></span><span class=line><span class=cl><span class=k>use</span> <span class=nn>DBI</span><span class=p>;</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nv>$dbName</span> <span class=o>=</span> <span class=s>&#39;bioinf&#39;</span><span class=p>;</span> 
</span></span><span class=line><span class=cl><span class=nv>$dbLogin</span> <span class=o>=</span> <span class=s>&#39;ezop&#39;</span><span class=p>;</span> 
</span></span><span class=line><span class=cl><span class=nv>$dbPass</span> <span class=o>=</span> <span class=s>&#39;XYZ&#39;</span><span class=p>;</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nv>$conn</span> <span class=o>=</span> <span class=nn>DBI</span><span class=o>-&gt;</span><span class=nb>connect</span><span class=p>(</span><span class=s>&#34;DBI:Pg:database=$dbName;host=localhost&#34;</span><span class=p>,</span> <span class=s>&#34;$dbLogin&#34;</span><span class=p>,</span> <span class=s>&#34;$dbPass&#34;</span><span class=p>,</span> <span class=p>{</span><span class=s>&#39;RaiseError&#39;</span> <span class=o>=&gt;</span> <span class=mi>1</span><span class=p>,</span> <span class=s>&#39;AutoCommit&#39;</span> <span class=o>=&gt;</span> <span class=mi>0</span><span class=p>});</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nb>die</span> <span class=s>&#34;./pdb_lines_unos.pl &lt;table&gt; &lt;file&gt;&#34;</span> <span class=k>if</span> <span class=ow>not</span> <span class=nb>defined</span><span class=p>(</span><span class=nv>$ARGV</span><span class=p>[</span><span class=mi>0</span><span class=p>]);</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nv>$recordCount</span> <span class=o>=</span> <span class=mi>0</span><span class=p>;</span>
</span></span><span class=line><span class=cl><span class=nv>$table</span> <span class=o>=</span> <span class=nv>$ARGV</span><span class=p>[</span><span class=mi>0</span><span class=p>];</span>
</span></span><span class=line><span class=cl><span class=nv>$fName</span> <span class=o>=</span> <span class=nv>$ARGV</span><span class=p>[</span><span class=mi>1</span><span class=p>];</span>
</span></span><span class=line><span class=cl><span class=nb>open</span> <span class=n>F</span><span class=p>,</span> <span class=s>&#34;zcat $fName|&#34;</span><span class=p>;</span>
</span></span><span class=line><span class=cl><span class=k>while</span> <span class=p>(</span><span class=sr>&lt;F&gt;</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=nb>chomp</span><span class=p>;</span>
</span></span><span class=line><span class=cl>    <span class=nv>$linija</span> <span class=o>=</span> <span class=nv>$_</span><span class=p>;</span>
</span></span><span class=line><span class=cl>    <span class=nv>$recordCount</span> <span class=o>+=</span> <span class=mi>1</span><span class=p>;</span>
</span></span><span class=line><span class=cl>    <span class=nv>$sql</span> <span class=o>=</span> <span class=s>&#34;insert into $table (code, line_num, line_cont) values (?, ?, ?)&#34;</span><span class=p>;</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>    <span class=nv>$conn</span><span class=o>-&gt;</span><span class=k>do</span><span class=p>(</span><span class=nv>$sql</span><span class=p>,</span> <span class=nb>undef</span><span class=p>,</span> <span class=nv>$fName</span><span class=p>,</span> <span class=nv>$recordCount</span><span class=p>,</span> <span class=nv>$linija</span><span class=p>);</span>
</span></span><span class=line><span class=cl>    <span class=nv>$conn</span><span class=o>-&gt;</span><span class=n>commit</span><span class=p>()</span> <span class=k>if</span> <span class=p>(</span><span class=nv>$recordCount%10000</span> <span class=o>==</span> <span class=mi>0</span><span class=p>);</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=nb>close</span> <span class=n>F</span><span class=p>;</span>
</span></span><span class=line><span class=cl><span class=nv>$conn</span><span class=o>-&gt;</span><span class=n>commit</span><span class=p>();</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=mi>1</span><span class=p>;</span>
</span></span></code></pre></td></tr></table></div></div><p>After importing lines, they are parsed using functions that will be defined later.</p><p>From the data in <code>raw_pdb</code>, we generate the <code>ts</code>, <code>ps</code>, <code>proteins</code>, <code>sources</code>, <code>sources_organela</code>, and <code>ss_bond</code> tables by parsing their corresponding records.</p><p>The <code>ps</code> table comprises three essential columns: <code>chain</code>, <code>length</code>, and <code>sequence</code>. The protein sequence is generated using C-alpha atoms for each chain and residue, ordered by residue sequence, considering only the first insertion and alternate location. The <code>chain</code> value is derived from the <code>TS.chain</code> column, while <code>length</code> represents the pre-calculated length of the <code>sequence</code> string. As this article focuses solely on single chains and intrachain connections, proteins with multiple chains are excluded from the analysis.</p><p>All disulfide bonds within <code>SSBOND</code> records are stored in the <code>pdb_ssbond</code> table, which inherits from the <code>pdb_ssbond_extended</code> table. The structure of <code>pdb_ssbond</code> is as follows:</p><div class=responsive-table><div class=table-wrapper><table><thead><tr><th>Column</th><th>Type</th><th>Nullable</th><th>Default</th><th>Description</th></tr></thead><tbody><tr><td>id</td><td>integer</td><td>not null</td><td>nextval(‘pdb_ssbond_id_seq’::regclass)</td><td>&nbsp;</td></tr><tr><td>code</td><td>character(4)</td><td>&nbsp;</td><td>&nbsp;</td><td>four-letter code</td></tr><tr><td>ser_num</td><td>integer</td><td>&nbsp;</td><td>&nbsp;</td><td>serial number of ssbond</td></tr><tr><td>residue1</td><td>character(3)</td><td>&nbsp;</td><td>&nbsp;</td><td>first residue in bond</td></tr><tr><td>chain_id1</td><td>character(1)</td><td>&nbsp;</td><td>&nbsp;</td><td>first chain in bond</td></tr><tr><td>res_seq1</td><td>integer</td><td>&nbsp;</td><td>&nbsp;</td><td>sequential number of first residue</td></tr><tr><td>i_code1</td><td>character(1)</td><td>&nbsp;</td><td>&nbsp;</td><td>insertion code of first residue</td></tr><tr><td>residue2</td><td>character(3)</td><td>&nbsp;</td><td>&nbsp;</td><td>second residue in bond</td></tr><tr><td>chain_id2</td><td>character(1)</td><td>&nbsp;</td><td>&nbsp;</td><td>second chain in bond</td></tr><tr><td>res_seq2</td><td>integer</td><td>&nbsp;</td><td>&nbsp;</td><td>sequential number of second residue</td></tr><tr><td>i_code2</td><td>character(1)</td><td>&nbsp;</td><td>&nbsp;</td><td>insertion code of second residue</td></tr><tr><td>sym1</td><td>character(6)</td><td>&nbsp;</td><td>&nbsp;</td><td>first symmetry operator</td></tr><tr><td>sym2</td><td>character(6)</td><td>&nbsp;</td><td>&nbsp;</td><td>second symmetry operator</td></tr><tr><td>dist</td><td>numeric(5,2)</td><td>&nbsp;</td><td>&nbsp;</td><td>distance between atoms</td></tr><tr><td>is_reactive</td><td>boolean</td><td>not null</td><td>false</td><td>mark for reactivity (to be set)</td></tr><tr><td>is_consecutive</td><td>boolean</td><td>not null</td><td>true</td><td>mark for consecutive bonds (to be set)</td></tr><tr><td>rep7</td><td>boolean</td><td>not null</td><td>false</td><td>mark for set-7 structures (to be set)</td></tr><tr><td>rep40</td><td>boolean</td><td>not null</td><td>false</td><td>mark for set-40 structures (to be set)</td></tr><tr><td>rep80</td><td>boolean</td><td>not null</td><td>false</td><td>mark for set-80 structures (to be set)</td></tr><tr><td>is_from_pdb</td><td>boolean</td><td>&nbsp;</td><td>true</td><td>is taken from PDB as source (to be set)</td></tr></tbody></table></div></div><p>I have also added the following indexes:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-sql data-lang=sql><span class=line><span class=cl><span class=s2>&#34;pdb_ssbond_pkey&#34;</span><span class=w> </span><span class=k>PRIMARY</span><span class=w> </span><span class=k>KEY</span><span class=p>,</span><span class=w> </span><span class=n>btree</span><span class=w> </span><span class=p>(</span><span class=n>id</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=s2>&#34;ndxcode1&#34;</span><span class=w> </span><span class=n>btree</span><span class=w> </span><span class=p>(</span><span class=n>code</span><span class=p>,</span><span class=w> </span><span class=n>chain_id1</span><span class=p>,</span><span class=w> </span><span class=n>res_seq1</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=s2>&#34;ndxcode2&#34;</span><span class=w> </span><span class=n>btree</span><span class=w> </span><span class=p>(</span><span class=n>code</span><span class=p>,</span><span class=w> </span><span class=n>chain_id2</span><span class=p>,</span><span class=w> </span><span class=n>res_seq2</span><span class=p>)</span><span class=w>
</span></span></span></code></pre></td></tr></table></div></div><p>Assuming a Gaussian distribution of disulfide bond lengths prior to cutoff (without formal testing using methods like the KS-test), standard deviations were calculated for each distance between cysteines within the same protein to determine the acceptable range of bond lengths and compare them against the cutoff. The cutoff was defined as the calculated mean plus or minus three standard deviations. To encompass more potential disulfide bonds not explicitly listed in the <code>SSBOND</code> rows of the PDB file, the range was extended. These bonds were identified by calculating distances between <code>ATOM</code> records. The chosen range for <code>ssbonds</code> falls between 1.6175344456264 and 2.48801947642267 Å, representing the mean (2.05) plus or minus four standard deviations:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt> 1
</span><span class=lnt> 2
</span><span class=lnt> 3
</span><span class=lnt> 4
</span><span class=lnt> 5
</span><span class=lnt> 6
</span><span class=lnt> 7
</span><span class=lnt> 8
</span><span class=lnt> 9
</span><span class=lnt>10
</span><span class=lnt>11
</span><span class=lnt>12
</span><span class=lnt>13
</span><span class=lnt>14
</span><span class=lnt>15
</span><span class=lnt>16
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-sql data-lang=sql><span class=line><span class=cl><span class=k>select</span><span class=w>                     </span><span class=k>count</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>     </span><span class=n>cnt</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>      </span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w>             </span><span class=k>as</span><span class=w> </span><span class=n>std_dev</span><span class=w>  
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>                     </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>avg_val</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>     </span><span class=o>-</span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>   </span><span class=n>left1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>      </span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>  </span><span class=n>right1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w> </span><span class=o>-</span><span class=mi>2</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>   </span><span class=n>left2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>  </span><span class=mi>2</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>  </span><span class=n>right2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w> </span><span class=o>-</span><span class=mi>3</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>   </span><span class=n>left3</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>  </span><span class=mi>3</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>  </span><span class=n>right3</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w> </span><span class=o>-</span><span class=mi>4</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>   </span><span class=n>left4</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>  </span><span class=mi>4</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=n>stddev</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=k>avg</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w>  </span><span class=n>right4</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>         </span><span class=k>min</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w>         </span><span class=k>max</span><span class=p>(</span><span class=n>dist</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond_tmp</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>where</span><span class=w> </span><span class=n>dist</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=mi>0</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>and</span><span class=w> </span><span class=n>dist</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=mi>20</span><span class=p>;</span><span class=w>
</span></span></span></code></pre></td></tr></table></div></div><p>While the <code>TS</code> table contains coordinates for all atoms, our analysis will focus solely on cysteines with their sulfur atom labeled as <code>" SG "</code>. To expedite the process, a separate staging table containing only <code>" SG "</code> sulfur atoms is created, reducing the number of records to be searched. By selecting only sulfurs, the number of combinations is significantly reduced from 122,761,100 (for all atoms) to 194,574. Within this self-joined table, distances are calculated using the Euclidean distance formula, and the results are imported into the <code>pdb_ssbond</code> table, but only if the distance falls within the predefined length range calculated earlier. This speed optimization aims to minimize the time required to rerun the entire process for verification purposes, keeping it within a one-day timeframe.</p><p>The following algorithm is employed to clean disulfide bonds:</p><ul><li>Delete bonds where both connection points refer to the same amino acid.</li><li>Delete bonds with lengths outside the range of 1.6175344456264 and 2.48801947642267 Å.</li><li>Remove insertions.</li><li>Remove bonds resulting from alternate atom locations, retaining only the first location.</li></ul><p>The code implementing this algorithm (taking the <code>pdb_ssbond</code> table as the first argument) is as follows:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt>  1
</span><span class=lnt>  2
</span><span class=lnt>  3
</span><span class=lnt>  4
</span><span class=lnt>  5
</span><span class=lnt>  6
</span><span class=lnt>  7
</span><span class=lnt>  8
</span><span class=lnt>  9
</span><span class=lnt> 10
</span><span class=lnt> 11
</span><span class=lnt> 12
</span><span class=lnt> 13
</span><span class=lnt> 14
</span><span class=lnt> 15
</span><span class=lnt> 16
</span><span class=lnt> 17
</span><span class=lnt> 18
</span><span class=lnt> 19
</span><span class=lnt> 20
</span><span class=lnt> 21
</span><span class=lnt> 22
</span><span class=lnt> 23
</span><span class=lnt> 24
</span><span class=lnt> 25
</span><span class=lnt> 26
</span><span class=lnt> 27
</span><span class=lnt> 28
</span><span class=lnt> 29
</span><span class=lnt> 30
</span><span class=lnt> 31
</span><span class=lnt> 32
</span><span class=lnt> 33
</span><span class=lnt> 34
</span><span class=lnt> 35
</span><span class=lnt> 36
</span><span class=lnt> 37
</span><span class=lnt> 38
</span><span class=lnt> 39
</span><span class=lnt> 40
</span><span class=lnt> 41
</span><span class=lnt> 42
</span><span class=lnt> 43
</span><span class=lnt> 44
</span><span class=lnt> 45
</span><span class=lnt> 46
</span><span class=lnt> 47
</span><span class=lnt> 48
</span><span class=lnt> 49
</span><span class=lnt> 50
</span><span class=lnt> 51
</span><span class=lnt> 52
</span><span class=lnt> 53
</span><span class=lnt> 54
</span><span class=lnt> 55
</span><span class=lnt> 56
</span><span class=lnt> 57
</span><span class=lnt> 58
</span><span class=lnt> 59
</span><span class=lnt> 60
</span><span class=lnt> 61
</span><span class=lnt> 62
</span><span class=lnt> 63
</span><span class=lnt> 64
</span><span class=lnt> 65
</span><span class=lnt> 66
</span><span class=lnt> 67
</span><span class=lnt> 68
</span><span class=lnt> 69
</span><span class=lnt> 70
</span><span class=lnt> 71
</span><span class=lnt> 72
</span><span class=lnt> 73
</span><span class=lnt> 74
</span><span class=lnt> 75
</span><span class=lnt> 76
</span><span class=lnt> 77
</span><span class=lnt> 78
</span><span class=lnt> 79
</span><span class=lnt> 80
</span><span class=lnt> 81
</span><span class=lnt> 82
</span><span class=lnt> 83
</span><span class=lnt> 84
</span><span class=lnt> 85
</span><span class=lnt> 86
</span><span class=lnt> 87
</span><span class=lnt> 88
</span><span class=lnt> 89
</span><span class=lnt> 90
</span><span class=lnt> 91
</span><span class=lnt> 92
</span><span class=lnt> 93
</span><span class=lnt> 94
</span><span class=lnt> 95
</span><span class=lnt> 96
</span><span class=lnt> 97
</span><span class=lnt> 98
</span><span class=lnt> 99
</span><span class=lnt>100
</span><span class=lnt>101
</span><span class=lnt>102
</span><span class=lnt>103
</span><span class=lnt>104
</span><span class=lnt>105
</span><span class=lnt>106
</span><span class=lnt>107
</span><span class=lnt>108
</span><span class=lnt>109
</span><span class=lnt>110
</span><span class=lnt>111
</span><span class=lnt>112
</span><span class=lnt>113
</span><span class=lnt>114
</span><span class=lnt>115
</span><span class=lnt>116
</span><span class=lnt>117
</span><span class=lnt>118
</span><span class=lnt>119
</span><span class=lnt>120
</span><span class=lnt>121
</span><span class=lnt>122
</span><span class=lnt>123
</span><span class=lnt>124
</span><span class=lnt>125
</span><span class=lnt>126
</span><span class=lnt>127
</span><span class=lnt>128
</span><span class=lnt>129
</span><span class=lnt>130
</span><span class=lnt>131
</span><span class=lnt>132
</span><span class=lnt>133
</span><span class=lnt>134
</span><span class=lnt>135
</span><span class=lnt>136
</span><span class=lnt>137
</span><span class=lnt>138
</span><span class=lnt>139
</span><span class=lnt>140
</span><span class=lnt>141
</span><span class=lnt>142
</span><span class=lnt>143
</span><span class=lnt>144
</span><span class=lnt>145
</span><span class=lnt>146
</span><span class=lnt>147
</span><span class=lnt>148
</span><span class=lnt>149
</span><span class=lnt>150
</span><span class=lnt>151
</span><span class=lnt>152
</span><span class=lnt>153
</span><span class=lnt>154
</span><span class=lnt>155
</span><span class=lnt>156
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-sql data-lang=sql><span class=line><span class=cl><span class=w> </span><span class=k>CREATE</span><span class=w> </span><span class=k>OR</span><span class=w> </span><span class=k>REPLACE</span><span class=w> </span><span class=k>FUNCTION</span><span class=w> </span><span class=n>pdb_ssbond_clean2</span><span class=p>(</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>clean_icodes</span><span class=w> </span><span class=nb>boolean</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>clean_altloc</span><span class=w> </span><span class=nb>boolean</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>mark_reactive</span><span class=w> </span><span class=nb>boolean</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>mark_consecutive</span><span class=w> </span><span class=nb>boolean</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>RETURNS</span><span class=w> </span><span class=n>void</span><span class=w> </span><span class=k>AS</span><span class=w> </span><span class=err>$</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>declare</span><span class=w> </span><span class=n>_res</span><span class=w> </span><span class=nb>integer</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>BEGIN</span><span class=w>  
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>delete</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> </span><span class=n>b</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>where</span><span class=w> </span><span class=k>exists</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>select</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>id</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> </span><span class=n>a</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>where</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>code</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>id</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>id</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>and</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=p>(</span><span class=w>   </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>res_seq1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>res_seq2</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>or</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=p>(</span><span class=w>   </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>res_seq2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>res_seq1</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>)</span><span class=w> </span><span class=p>;</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>delete</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>where</span><span class=w> </span><span class=n>chain_id1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>chain_id2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>and</span><span class=w> </span><span class=n>res_seq1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>res_seq2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>and</span><span class=w> </span><span class=n>i_code1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>i_code2</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        
</span></span></span><span class=line><span class=cl><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>         
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>update</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>set</span><span class=w> </span><span class=n>is_consecutive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>true</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>,</span><span class=w> </span><span class=n>is_reactive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>false</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>delete</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>where</span><span class=w> </span><span class=k>not</span><span class=w> </span><span class=n>dist</span><span class=w> </span><span class=k>between</span><span class=w> </span><span class=mi>1</span><span class=p>.</span><span class=mi>6175344456264</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=mi>2</span><span class=p>.</span><span class=mi>48801947642267</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>     
</span></span></span><span class=line><span class=cl><span class=w>          
</span></span></span><span class=line><span class=cl><span class=w>    
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>if</span><span class=w> </span><span class=n>clean_icodes</span><span class=w> </span><span class=k>then</span><span class=w>  
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>delete</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>where</span><span class=w>  </span><span class=n>i_code1</span><span class=w> </span><span class=k>not</span><span class=w> </span><span class=k>in</span><span class=w> </span><span class=p>(</span><span class=s1>&#39;&#39;</span><span class=p>,</span><span class=w>  </span><span class=s1>&#39; &#39;</span><span class=p>,</span><span class=w> </span><span class=s1>&#39;A&#39;</span><span class=p>)</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>or</span><span class=w>     </span><span class=n>i_code2</span><span class=w> </span><span class=k>not</span><span class=w> </span><span class=k>in</span><span class=w> </span><span class=p>(</span><span class=s1>&#39;&#39;</span><span class=p>,</span><span class=w>  </span><span class=s1>&#39; &#39;</span><span class=p>,</span><span class=w> </span><span class=s1>&#39;A&#39;</span><span class=p>)</span><span class=w> </span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>end</span><span class=w> </span><span class=k>if</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>if</span><span class=w> </span><span class=n>clean_altloc</span><span class=w> </span><span class=k>then</span><span class=w>  
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>delete</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> </span><span class=n>a</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>where</span><span class=w> </span><span class=k>exists</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>select</span><span class=w> </span><span class=mi>1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>  </span><span class=n>b</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>where</span><span class=w>   </span><span class=n>b</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>code</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>i_code1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>i_code1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>ser_num</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>ser_num</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=p>)</span><span class=w> </span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>delete</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> </span><span class=n>a</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>where</span><span class=w> </span><span class=k>exists</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>select</span><span class=w> </span><span class=mi>1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>  </span><span class=n>b</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>where</span><span class=w>   </span><span class=n>b</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>code</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>res_seq2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>i_code2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>i_code2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>ser_num</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>ser_num</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=p>)</span><span class=w> </span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>end</span><span class=w> </span><span class=k>if</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>if</span><span class=w> </span><span class=n>mark_reactive</span><span class=w> </span><span class=k>then</span><span class=w>  
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>update</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>set</span><span class=w> </span><span class=n>is_reactive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>false</span><span class=w> </span><span class=p>;</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>       
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>update</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>set</span><span class=w> </span><span class=n>is_reactive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>true</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>where</span><span class=w> </span><span class=k>exists</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>select</span><span class=w> </span><span class=mi>1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>  </span><span class=n>b</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>where</span><span class=w>   </span><span class=n>b</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>code</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>chain_id1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>i_code1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>i_code1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>ser_num</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>ser_num</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=p>)</span><span class=w> </span><span class=p>;</span><span class=w>  
</span></span></span><span class=line><span class=cl><span class=w>         
</span></span></span><span class=line><span class=cl><span class=w>       
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>update</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>set</span><span class=w> </span><span class=n>is_reactive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>true</span><span class=w>         
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>where</span><span class=w> </span><span class=k>exists</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>select</span><span class=w> </span><span class=mi>1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>  </span><span class=n>b</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>where</span><span class=w>   </span><span class=n>b</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>code</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>chain_id2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>res_seq2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>i_code2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>i_code2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=n>b</span><span class=p>.</span><span class=n>ser_num</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>ser_num</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=p>)</span><span class=w> </span><span class=p>;</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>       
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>update</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>set</span><span class=w> </span><span class=n>is_reactive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>true</span><span class=w>         
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>where</span><span class=w> </span><span class=p>(</span><span class=n>code</span><span class=p>,</span><span class=w> </span><span class=n>chain_id1</span><span class=p>,</span><span class=w> </span><span class=n>res_seq1</span><span class=p>,</span><span class=w> </span><span class=n>i_code1</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>in</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>select</span><span class=w> </span><span class=n>code</span><span class=p>,</span><span class=w> </span><span class=n>chain_id1</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=k>c</span><span class=p>,</span><span class=w> </span><span class=n>res_seq1</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>r</span><span class=p>,</span><span class=w> </span><span class=n>i_code1</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>i</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>intersect</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>select</span><span class=w> </span><span class=n>code</span><span class=p>,</span><span class=w> </span><span class=n>chain_id2</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=k>c</span><span class=p>,</span><span class=w> </span><span class=n>res_seq2</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>r</span><span class=p>,</span><span class=w> </span><span class=n>i_code2</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>i</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=p>)</span><span class=w> </span><span class=p>;</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>       
</span></span></span><span class=line><span class=cl><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>update</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>set</span><span class=w> </span><span class=n>is_reactive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>true</span><span class=w>         
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>where</span><span class=w> </span><span class=p>(</span><span class=n>code</span><span class=p>,</span><span class=w> </span><span class=n>chain_id2</span><span class=p>,</span><span class=w> </span><span class=n>res_seq2</span><span class=p>,</span><span class=w> </span><span class=n>i_code2</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>in</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>select</span><span class=w> </span><span class=n>code</span><span class=p>,</span><span class=w> </span><span class=n>chain_id1</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=k>c</span><span class=p>,</span><span class=w> </span><span class=n>res_seq1</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>r</span><span class=p>,</span><span class=w> </span><span class=n>i_code1</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>i</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>intersect</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>select</span><span class=w> </span><span class=n>code</span><span class=p>,</span><span class=w> </span><span class=n>chain_id2</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=k>c</span><span class=p>,</span><span class=w> </span><span class=n>res_seq2</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>r</span><span class=p>,</span><span class=w> </span><span class=n>i_code2</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>i</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=p>);</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>end</span><span class=w> </span><span class=k>if</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>if</span><span class=w> </span><span class=n>mark_consecutive</span><span class=w> </span><span class=k>then</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>           </span><span class=k>update</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>           </span><span class=k>set</span><span class=w> </span><span class=n>is_consecutive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>false</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>update</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>set</span><span class=w> </span><span class=n>is_consecutive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>true</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=k>where</span><span class=w> </span><span class=k>not</span><span class=w> </span><span class=k>exists</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>select</span><span class=w> </span><span class=mi>1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>from</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=w> </span><span class=n>a</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                </span><span class=k>where</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=n>a</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>code</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=p>(</span><span class=n>a</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>chain_id2</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=k>or</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=p>(</span><span class=n>a</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>chain_id1</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                           </span><span class=p>(</span><span class=n>a</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=k>between</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq2</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=k>or</span><span class=w> </span><span class=p>(</span><span class=n>a</span><span class=p>.</span><span class=n>res_seq2</span><span class=w> </span><span class=k>between</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq2</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=k>or</span><span class=w> </span><span class=p>(</span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=k>between</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq2</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=k>or</span><span class=w> </span><span class=p>(</span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq2</span><span class=w> </span><span class=k>between</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq2</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=k>and</span><span class=w> </span><span class=k>not</span><span class=w> </span><span class=p>(</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                            </span><span class=n>a</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>code</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>chain_id1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>chain_id2</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                        </span><span class=k>and</span><span class=w> </span><span class=n>a</span><span class=p>.</span><span class=n>res_seq2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pdb_ssbond</span><span class=p>.</span><span class=n>res_seq2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                    </span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>            </span><span class=p>);</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>end</span><span class=w> </span><span class=k>if</span><span class=p>;</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>END</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=err>$</span><span class=w> </span><span class=k>LANGUAGE</span><span class=w> </span><span class=n>plpgsql</span><span class=p>;</span><span class=w>
</span></span></span></code></pre></td></tr></table></div></div><p>Subsequently, the non-redundant set of proteins is imported into the <code>nr</code> table, which is then joined with the <code>ps</code> and <code>proteins</code> tables. Sets are marked using <code>set7</code>, <code>set40</code>, and <code>set80</code>. Ultimately, only one set will be analyzed based on protein quantity. Any mismatched chains between PDB and NR are excluded from the analysis.</p><p>Proteins lacking disulfide bonds are removed from the study, along with those not belonging to any set. Data is processed using DSSP, and proteins exhibiting resolution issues or an excessive number of atoms are also excluded. Only proteins with single chains are retained for analysis, as interchain connections are not maintained, although they can be easily calculated from the <code>ssbond</code> table by counting connections with different chains.</p><p>For the remaining proteins, the total number of bonds and overlapping bonds are updated for each set.</p><p>The source organism is extracted from <code>SOURCE</code> records. Entries marked as unknown, synthetic, designed, artificial, or hybrid are discarded from the study. Low-resolution proteins are excluded only if their side chains are not visible.</p><p><code>SOURCE</code> records are stored in the <code>sources</code> table, which includes taxonomy rows. In cases of missing or incorrect taxonomy information, manual correction by experts is required.</p><p>Based on the source and taxonomy information downloaded from NCBI, each primary structure is assigned a class. Proteins with unassigned classes are removed from the analysis list. Due to the reliance on biological databases, it is highly recommended that a biologist performs an additional verification of all source records and NCBI taxonomy classifications to avoid potential misclassifications between different domains.</p><p>To map source cellular locations to taxonomy IDs, data from the source table is extracted into the <code>sources_organela</code> table, where all data is represented using codes, tags, and values. Its format is shown below:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt>1
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-sql data-lang=sql><span class=line><span class=cl><span class=k>select</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>sources_organela</span><span class=w> </span><span class=k>where</span><span class=w> </span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s1>&#39;1rav&#39;</span><span class=p>;</span><span class=w>
</span></span></span></code></pre></td></tr></table></div></div><div class=responsive-table><div class=table-wrapper><table><thead><tr><th>code</th><th>mol_id</th><th>tag</th><th>val</th></tr></thead><tbody><tr><td>1rav</td><td>0</td><td>MOL_ID</td><td>1</td></tr><tr><td>1rav</td><td>7</td><td>CELLULAR_LOCATION</td><td>CYTOPLASM (WHITE)</td></tr></tbody></table></div></div><p>The taxonomy archive used in this study is a ZIP file containing seven dump files. Among these, <code>names.dmp</code> and <code>merged.dmp</code> are particularly important. Both are CSV files using tab and pipe delimiters as <a class=link href=https://ftp.ncbi.nih.gov/pub/taxonomy/taxdump_readme.txt target=_blank rel=noopener>detailed in the documentation</a>:</p><ul><li><code>merged.dmp</code> provides a mapping of previous taxonomy IDs to their current counterparts.</li><li><code>names.dmp</code> contains the following essential columns in the specified order:<ul><li><code>tax_id</code>: The taxonomy ID.</li><li><code>name_txt</code>: The species name and, if applicable, the unique name (used for disambiguation when multiple names exist for a species).</li></ul></li><li><code>division.dmp</code> contains the names of top-level domains, which will serve as our classes.</li><li><code>nodes.dmp</code> represents the hierarchical structure of organisms based on taxonomy IDs.<ul><li>It includes a parent taxonomy ID, serving as a foreign key referencing <code>names.dmp</code>.</li><li>It also contains a division ID, crucial for storing relevant top-domain data.</li></ul></li></ul><p>Using this data and manual corrections (to ensure accurate domains of life assignments), we constructed the <code>taxonomy_path</code> table. A sample of the data is shown below:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt>1
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-sql data-lang=sql><span class=line><span class=cl><span class=k>select</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>taxonomy_path</span><span class=w> </span><span class=k>order</span><span class=w> </span><span class=k>by</span><span class=w> </span><span class=k>length</span><span class=p>(</span><span class=n>path</span><span class=p>)</span><span class=w>  </span><span class=k>limit</span><span class=w> </span><span class=mi>20</span><span class=w> </span><span class=k>offset</span><span class=w> </span><span class=mi>2000</span><span class=p>;</span><span class=w>
</span></span></span></code></pre></td></tr></table></div></div><div class=responsive-table><div class=table-wrapper><table><thead><tr><th>tax_id</th><th>path</th><th>is_viral</th><th>is_eukaryote</th><th>is_archaea</th><th>is_other</th><th>is_prokaryote</th></tr></thead><tbody><tr><td>142182</td><td>cellular organisms;Bacteria;Gemmatimonadetes</td><td>f</td><td>f</td><td>f</td><td>f</td><td>t</td></tr><tr><td>136087</td><td>cellular organisms;Eukaryota;Malawimonadidae</td><td>f</td><td>t</td><td>f</td><td>f</td><td>f</td></tr><tr><td>649454</td><td>Viruses;unclassified phages;Cyanophage G1168</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>321302</td><td>Viruses;unclassified viruses;Tellina virus 1</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>649453</td><td>Viruses;unclassified phages;Cyanophage G1158</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>536461</td><td>Viruses;unclassified phages;Cyanophage S-SM1</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>536462</td><td>Viruses;unclassified phages;Cyanophage S-SM2</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>77041</td><td>Viruses;unclassified viruses;Stealth virus 4</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>77042</td><td>Viruses;unclassified viruses;Stealth virus 5</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>641835</td><td>Viruses;unclassified phages;Vibrio phage 512</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>1074427</td><td>Viruses;unclassified viruses;Mouse Rosavirus</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>1074428</td><td>Viruses;unclassified viruses;Mouse Mosavirus</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>480920</td><td>other sequences;plasmids;IncP-1 plasmid 6-S1</td><td>f</td><td>f</td><td>f</td><td>t</td><td>f</td></tr><tr><td>2441</td><td>other sequences;plasmids;Plasmid ColBM-Cl139</td><td>f</td><td>f</td><td>f</td><td>t</td><td>f</td></tr><tr><td>168317</td><td>other sequences;plasmids;IncQ plasmid pIE723</td><td>f</td><td>f</td><td>f</td><td>t</td><td>f</td></tr><tr><td>536472</td><td>Viruses;unclassified phages;Cyanophage Syn10</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>536474</td><td>Viruses;unclassified phages;Cyanophage Syn30</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>2407</td><td>other sequences;transposons;Transposon Tn501</td><td>f</td><td>f</td><td>f</td><td>t</td><td>f</td></tr><tr><td>227307</td><td>Viruses;ssDNA viruses;Circoviridae;Gyrovirus</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr><tr><td>687247</td><td>Viruses;unclassified phages;Cyanophage ZQS-7</td><td>t</td><td>f</td><td>f</td><td>f</td><td>f</td></tr></tbody></table></div></div><p>To mitigate biases, sequences must undergo identity level checks before analysis. Although the NR set contains pre-compared sequences, an additional verification is always recommended.</p><p>For statistical analysis of each disulfide bond, data is tagged as reactive or overlapping. By marking overlaps, we automatically determine the number of consecutive and non-consecutive bonds within each protein. This information is stored in the <code>proteins</code> table, from which all protein complexes are excluded in the final results.</p><p>Each disulfide bond is also associated with sets by checking if both bond sides belong to the same NR set. Bonds with sides belonging to different sets are omitted from the analysis for that specific set.</p><p>Analyzing the quantity of bonds based on their variance requires categorizing each length into a specific class. In this study, we used five classes, as defined in the function below:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt> 1
</span><span class=lnt> 2
</span><span class=lnt> 3
</span><span class=lnt> 4
</span><span class=lnt> 5
</span><span class=lnt> 6
</span><span class=lnt> 7
</span><span class=lnt> 8
</span><span class=lnt> 9
</span><span class=lnt>10
</span><span class=lnt>11
</span><span class=lnt>12
</span><span class=lnt>13
</span><span class=lnt>14
</span><span class=lnt>15
</span><span class=lnt>16
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-sql data-lang=sql><span class=line><span class=cl><span class=k>CREATE</span><span class=w> </span><span class=k>OR</span><span class=w> </span><span class=k>REPLACE</span><span class=w> </span><span class=k>FUNCTION</span><span class=w> </span><span class=n>len2class</span><span class=p>(</span><span class=n>len</span><span class=w> </span><span class=nb>integer</span><span class=p>)</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>RETURNS</span><span class=w> </span><span class=nb>integer</span><span class=w> </span><span class=k>AS</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=err>$</span><span class=n>BODY$</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>BEGIN</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>return</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>case</span><span class=w> 
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>when</span><span class=w> </span><span class=n>len</span><span class=w> </span><span class=o>&lt;=</span><span class=w> </span><span class=mi>100</span><span class=w>                </span><span class=k>then</span><span class=w> </span><span class=mi>1</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>when</span><span class=w> </span><span class=n>len</span><span class=w> </span><span class=o>&gt;</span><span class=w>  </span><span class=mi>100</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>len</span><span class=w> </span><span class=o>&lt;=</span><span class=w> </span><span class=mi>200</span><span class=w> </span><span class=k>then</span><span class=w> </span><span class=mi>2</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>when</span><span class=w> </span><span class=n>len</span><span class=w> </span><span class=o>&gt;</span><span class=w>  </span><span class=mi>200</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>len</span><span class=w> </span><span class=o>&lt;=</span><span class=w> </span><span class=mi>300</span><span class=w> </span><span class=k>then</span><span class=w> </span><span class=mi>3</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>        </span><span class=k>when</span><span class=w> </span><span class=n>len</span><span class=w> </span><span class=o>&gt;</span><span class=w>  </span><span class=mi>300</span><span class=w> </span><span class=k>and</span><span class=w> </span><span class=n>len</span><span class=w> </span><span class=o>&lt;=</span><span class=w> </span><span class=mi>400</span><span class=w> </span><span class=k>then</span><span class=w> </span><span class=mi>4</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>                                       </span><span class=k>else</span><span class=w> </span><span class=mi>5</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>end</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=k>END</span><span class=p>;</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w></span><span class=err>$</span><span class=n>BODY$</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=k>LANGUAGE</span><span class=w> </span><span class=n>plpgsql</span><span class=w> </span><span class=k>VOLATILE</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>COST</span><span class=w> </span><span class=mi>100</span><span class=p>;</span><span class=w>
</span></span></span></code></pre></td></tr></table></div></div><p>Since most protein sizes are below 400 amino acids, length classification is performed by dividing lengths into five ranges. The first four ranges span 100 amino acids each, while the last range encompasses the remaining lengths. The following code snippet demonstrates how to utilize this function for data extraction and analysis:</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre tabindex=0 class=chroma><code><span class=lnt> 1
</span><span class=lnt> 2
</span><span class=lnt> 3
</span><span class=lnt> 4
</span><span class=lnt> 5
</span><span class=lnt> 6
</span><span class=lnt> 7
</span><span class=lnt> 8
</span><span class=lnt> 9
</span><span class=lnt>10
</span><span class=lnt>11
</span><span class=lnt>12
</span><span class=lnt>13
</span><span class=lnt>14
</span><span class=lnt>15
</span><span class=lnt>16
</span><span class=lnt>17
</span><span class=lnt>18
</span><span class=lnt>19
</span><span class=lnt>20
</span><span class=lnt>21
</span><span class=lnt>22
</span><span class=lnt>23
</span><span class=lnt>24
</span><span class=lnt>25
</span><span class=lnt>26
</span><span class=lnt>27
</span><span class=lnt>28
</span><span class=lnt>29
</span><span class=lnt>30
</span><span class=lnt>31
</span></code></pre></td><td class=lntd><pre tabindex=0 class=chroma><code class=language-sql data-lang=sql><span class=line><span class=cl><span class=w> </span><span class=k>SELECT</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>code</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>title</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>ss_bonds</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>ssbonds_overlap</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>intra_count</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>sci_name_src</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>len</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>tax_path</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>pfam_families</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>src_class</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>(</span><span class=w> </span><span class=k>SELECT</span><span class=w> </span><span class=n>s</span><span class=p>.</span><span class=n>id</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>           </span><span class=k>FROM</span><span class=w> </span><span class=n>src_classes</span><span class=w> </span><span class=n>s</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>          </span><span class=k>WHERE</span><span class=w> </span><span class=n>s</span><span class=p>.</span><span class=n>src_class</span><span class=p>::</span><span class=nb>text</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>src_class</span><span class=p>::</span><span class=nb>text</span><span class=p>)</span><span class=w> </span><span class=k>AS</span><span class=w> </span><span class=n>src_class_id</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>p</span><span class=p>.</span><span class=n>len_class7</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>(</span><span class=w> </span><span class=k>SELECT</span><span class=w> </span><span class=n>s</span><span class=p>.</span><span class=n>val</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>           </span><span class=k>FROM</span><span class=w> </span><span class=n>sources_organela</span><span class=w> </span><span class=n>s</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>          </span><span class=k>WHERE</span><span class=w> </span><span class=n>s</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>code</span><span class=p>::</span><span class=n>bpchar</span><span class=w> </span><span class=k>AND</span><span class=w> </span><span class=n>s</span><span class=p>.</span><span class=n>tag</span><span class=p>::</span><span class=nb>text</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s1>&#39;EXPRESSION_SYSTEM_CELLULAR_LOCATION&#39;</span><span class=p>::</span><span class=nb>text</span><span class=p>)</span><span class=w> </span><span class=k>AS</span><span class=w> </span><span class=n>expression_system_cellular_location</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=p>(</span><span class=w> </span><span class=k>SELECT</span><span class=w> </span><span class=n>s</span><span class=p>.</span><span class=n>val</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>           </span><span class=k>FROM</span><span class=w> </span><span class=n>sources_organela</span><span class=w> </span><span class=n>s</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>          </span><span class=k>WHERE</span><span class=w> </span><span class=n>s</span><span class=p>.</span><span class=n>code</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>code</span><span class=p>::</span><span class=n>bpchar</span><span class=w> </span><span class=k>AND</span><span class=w> </span><span class=n>s</span><span class=p>.</span><span class=n>tag</span><span class=p>::</span><span class=nb>text</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s1>&#39;CELLULAR_LOCATION&#39;</span><span class=p>::</span><span class=nb>text</span><span class=p>)</span><span class=w> </span><span class=k>AS</span><span class=w> </span><span class=n>cellular_location</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>ps</span><span class=p>.</span><span class=n>sequence</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>ps</span><span class=p>.</span><span class=n>uniprot_code</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>ps</span><span class=p>.</span><span class=n>accession_code</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>ps</span><span class=p>.</span><span class=n>cc</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>ps</span><span class=p>.</span><span class=n>seq_uniprot</span><span class=p>,</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>    </span><span class=n>ps</span><span class=p>.</span><span class=n>chain_id</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>   </span><span class=k>FROM</span><span class=w> </span><span class=n>proteins</span><span class=w> </span><span class=n>p</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>     </span><span class=k>JOIN</span><span class=w> </span><span class=n>nr</span><span class=w> </span><span class=n>n</span><span class=w> </span><span class=k>ON</span><span class=w> </span><span class=n>n</span><span class=p>.</span><span class=n>code</span><span class=p>::</span><span class=nb>text</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>code</span><span class=p>::</span><span class=nb>text</span><span class=w> </span><span class=k>AND</span><span class=w> </span><span class=n>n</span><span class=p>.</span><span class=n>rep7</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>true</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>     </span><span class=k>JOIN</span><span class=w> </span><span class=n>ps</span><span class=w> </span><span class=n>ps</span><span class=w> </span><span class=k>ON</span><span class=w> </span><span class=n>ps</span><span class=p>.</span><span class=n>code</span><span class=p>::</span><span class=nb>text</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>n</span><span class=p>.</span><span class=n>code</span><span class=p>::</span><span class=nb>text</span><span class=w> </span><span class=k>AND</span><span class=w> </span><span class=n>ps</span><span class=p>.</span><span class=n>chain_id</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>n</span><span class=p>.</span><span class=n>chain_id</span><span class=w> </span><span class=k>AND</span><span class=w> </span><span class=n>ps</span><span class=p>.</span><span class=n>het</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>false</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>  </span><span class=k>WHERE</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>is_excluded</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>false</span><span class=w> </span><span class=k>AND</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>chain_cnt</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=mi>1</span><span class=w> </span><span class=k>AND</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>is_set7</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=k>true</span><span class=w> </span><span class=k>AND</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>reactive_cnt7</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=mi>0</span><span class=w>
</span></span></span><span class=line><span class=cl><span class=w>  </span><span class=k>ORDER</span><span class=w> </span><span class=k>BY</span><span class=w> </span><span class=n>p</span><span class=p>.</span><span class=n>code</span><span class=p>;</span><span class=w>
</span></span></span></code></pre></td></tr></table></div></div><p>An example of the output, incorporating corrected titles and manually added columns, is provided below:</p><div class=responsive-table><div class=table-wrapper><table><thead><tr><th>PDB code</th><th>Total number of SS-bonds</th><th>Number of non-consecutive SS-bonds</th><th>PDB length / amino acids</th><th>Domain</th><th>TargetP 1.1</th><th>TatP 1.0</th><th>SignalP 4.1</th><th>ChloroP 1.1</th><th>TMHMM 2.0 number of transmembrane domains</th><th>Big-pi</th><th>nucPred</th><th>NetNES 1.1</th><th>PSORTb v3.0</th><th>SecretomeP 2.0</th><th>LocTree2</th><th>Consensus localization prediction</th></tr></thead><tbody><tr><td>1akp</td><td>2</td><td>0</td><td>114</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>no signal peptide</td><td>ND</td><td>0</td><td>ND</td><td>ND</td><td>ND</td><td>unknown</td><td>ND</td><td>fimbrium</td><td>unknown</td></tr><tr><td>1bhu</td><td>2</td><td>0</td><td>102</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>signal peptide</td><td>ND</td><td>1</td><td>ND</td><td>ND</td><td>ND</td><td>unknown</td><td>ND</td><td>secreted</td><td>unknown</td></tr><tr><td>1c75</td><td>0</td><td>0</td><td>71</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>no signal peptide</td><td>ND</td><td>0</td><td>ND</td><td>ND</td><td>ND</td><td>cytoplasmic membrane</td><td>nonclassical secretion</td><td>periplasm</td><td>unknown</td></tr><tr><td>1c8x</td><td>0</td><td>0</td><td>265</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>signal peptide</td><td>ND</td><td>1</td><td>ND</td><td>ND</td><td>ND</td><td>unknown</td><td>ND</td><td>secreted</td><td>unknown</td></tr><tr><td>1cx1</td><td>1</td><td>0</td><td>153</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>signal peptide</td><td>ND</td><td>1</td><td>ND</td><td>ND</td><td>ND</td><td>extracellular</td><td>ND</td><td>secreted</td><td>unknown</td></tr><tr><td>1dab</td><td>0</td><td>0</td><td>539</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>signal peptide</td><td>ND</td><td>0</td><td>ND</td><td>ND</td><td>ND</td><td>outer membrane</td><td>ND</td><td>outer membrane</td><td>unknown</td></tr><tr><td>1dfu</td><td>0</td><td>0</td><td>94</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>no signal peptide</td><td>ND</td><td>0</td><td>ND</td><td>ND</td><td>ND</td><td>cytoplasmic</td><td>ND</td><td>cytosol</td><td>unknown</td></tr><tr><td>1e8r</td><td>2</td><td>2</td><td>50</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>signal peptide</td><td>ND</td><td>0</td><td>ND</td><td>ND</td><td>ND</td><td>unknown</td><td>ND</td><td>secreted</td><td>secreted</td></tr><tr><td>1esc</td><td>3</td><td>0</td><td>302</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>signal peptide</td><td>ND</td><td>1</td><td>ND</td><td>ND</td><td>ND</td><td>extracellular</td><td>ND</td><td>periplasm</td><td>unknown</td></tr><tr><td>1g6e</td><td>1</td><td>0</td><td>87</td><td>Bacteria</td><td>ND</td><td>Tat-signal</td><td>signal peptide</td><td>ND</td><td>1</td><td>ND</td><td>ND</td><td>ND</td><td>unknown</td><td>ND</td><td>secreted</td><td>unknown</td></tr></tbody></table></div></div><h2 id=postgresql-as-a-processing-intermediary>PostgreSQL as a Processing Intermediary</h2><p>This work outlined a comprehensive data processing pipeline, from fetching to analysis. In scientific data analysis, normalization might not always be necessary. When dealing with small datasets intended for one-time analysis, denormalization can be sufficient if processing speed is adequate.</p><p>The decision to perform all processing within a single bioinformatics database stems from PostgreSQL&rsquo;s ability to integrate with various languages, including R. This enables statistical analysis directly within the database—a topic for a future article on bioinformatics tools.</p><hr><p><em>Special thanks to my Toptal colleagues Stefan Fuchs and Aldo Zelen for their invaluable insights and consultations.</em></p></section><footer class=article-footer><section class=article-copyright><svg class="icon icon-tabler icon-tabler-copyright" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="12" cy="12" r="9"/><path d="M14.5 9a3.5 4 0 100 6"/></svg>
<span>Licensed under CC BY-NC-SA 4.0</span></section></footer></article><footer class=site-footer><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-3987358164777632" crossorigin=anonymous></script><script>(function(e,t,n,s,o){e[s]=e[s]||[],e[s].push({"gtm.start":(new Date).getTime(),event:"gtm.js"});var a=t.getElementsByTagName(n)[0],i=t.createElement(n),r=s!="dataLayer"?"&l="+s:"";i.async=!0,i.src="https://www.googletagmanager.com/gtm.js?id="+o+r,a.parentNode.insertBefore(i,a)})(window,document,"script","dataLayer","GTM-5X67X4Q4")</script><noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-5X67X4Q4" height=0 width=0 style=display:none;visibility:hidden></iframe></noscript></footer><div class=pswp tabindex=-1 role=dialog aria-hidden=true><div class=pswp__bg></div><div class=pswp__scroll-wrap><div class=pswp__container><div class=pswp__item></div><div class=pswp__item></div><div class=pswp__item></div></div><div class="pswp__ui pswp__ui--hidden"><div class=pswp__top-bar><div class=pswp__counter></div><button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
<button class="pswp__button pswp__button--share" title=Share></button>
<button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
<button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button><div class=pswp__preloader><div class=pswp__preloader__icn><div class=pswp__preloader__cut><div class=pswp__preloader__donut></div></div></div></div></div><div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap"><div class=pswp__share-tooltip></div></div><button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
</button>
<button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)"></button><div class=pswp__caption><div class=pswp__caption__center></div></div></div></div></div><script src=https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.js integrity="sha256-ePwmChbbvXbsO02lbM3HoHbSHTHFAeChekF1xKJdleo=" crossorigin=anonymous defer></script><script src=https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe-ui-default.min.js integrity="sha256-UKkzOn/w1mBxRmLLGrSeyB4e1xbrp4xylgAWb3M42pU=" crossorigin=anonymous defer></script><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/default-skin/default-skin.min.css crossorigin=anonymous><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.css crossorigin=anonymous></main></div><script src=https://cdn.jsdelivr.net/npm/node-vibrant@3.1.6/dist/vibrant.min.js integrity="sha256-awcR2jno4kI5X0zL8ex0vi2z+KMkF24hUW8WePSA9HM=" crossorigin=anonymous></script><script type=text/javascript src=/ts/main.js defer></script><script>(function(){const e=document.createElement("link");e.href="https://fonts.googleapis.com/css2?family=Lato:wght@300;400;700&display=swap",e.type="text/css",e.rel="stylesheet",document.head.appendChild(e)})()</script></body></html>