<!doctype html><html lang=en-us dir=ltr><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Archaeal Richmond Mine acidophilic nanoorganisms (ARMAN) were first discovered in an extremely acidic mine located in northern California (Richmond Mine at Iron Mountain) by Brett Baker in Jill Banfield&amp;rsquo;s laboratory at the University of California Berkeley. These novel groups of archaea named ARMAN-1, ARMAN-2 (Candidatus Micrarchaeum acidiphilum ARMAN-2 ), and ARMAN-3 were missed by previous PCR-based surveys of the mine community because the ARMANs have several mismatches with commonly used PCR primers for 16S rRNA genes."><title>Archaeal Richmond Mine Acidophilic Nanoorganisms</title><link rel=canonical href=https://Nexus-Security.github.io/posts/2016-09-09-archaeal-richmond-mine-acidophilic/><link rel=stylesheet href=/scss/style.min.450926226e724574a6b936335ea06111f8aeb253d932c86cb2cc807341cd2889.css><meta property="og:title" content="Archaeal Richmond Mine Acidophilic Nanoorganisms"><meta property="og:description" content="Archaeal Richmond Mine acidophilic nanoorganisms (ARMAN) were first discovered in an extremely acidic mine located in northern California (Richmond Mine at Iron Mountain) by Brett Baker in Jill Banfield&amp;rsquo;s laboratory at the University of California Berkeley. These novel groups of archaea named ARMAN-1, ARMAN-2 (Candidatus Micrarchaeum acidiphilum ARMAN-2 ), and ARMAN-3 were missed by previous PCR-based surveys of the mine community because the ARMANs have several mismatches with commonly used PCR primers for 16S rRNA genes."><meta property="og:url" content="https://Nexus-Security.github.io/posts/2016-09-09-archaeal-richmond-mine-acidophilic/"><meta property="og:site_name" content="ZYChimne"><meta property="og:type" content="article"><meta property="article:section" content="Posts"><meta property="article:published_time" content="2020-01-13T07:47:00+01:00"><meta property="article:modified_time" content="2020-01-13T07:47:00+01:00"><meta name=twitter:title content="Archaeal Richmond Mine Acidophilic Nanoorganisms"><meta name=twitter:description content="Archaeal Richmond Mine acidophilic nanoorganisms (ARMAN) were first discovered in an extremely acidic mine located in northern California (Richmond Mine at Iron Mountain) by Brett Baker in Jill Banfield&amp;rsquo;s laboratory at the University of California Berkeley. These novel groups of archaea named ARMAN-1, ARMAN-2 (Candidatus Micrarchaeum acidiphilum ARMAN-2 ), and ARMAN-3 were missed by previous PCR-based surveys of the mine community because the ARMANs have several mismatches with commonly used PCR primers for 16S rRNA genes."></head><body class=article-page><script>(function(){const e="StackColorScheme";localStorage.getItem(e)||localStorage.setItem(e,"auto")})()</script><script>(function(){const t="StackColorScheme",e=localStorage.getItem(t),n=window.matchMedia("(prefers-color-scheme: dark)").matches===!0;e=="dark"||e==="auto"&&n?document.documentElement.dataset.scheme="dark":document.documentElement.dataset.scheme="light"})()</script><div class="container main-container flex on-phone--column extended"><aside class="sidebar left-sidebar sticky"><button class="hamburger hamburger--spin" type=button id=toggle-menu aria-label="Toggle Menu">
<span class=hamburger-box><span class=hamburger-inner></span></span></button><header><figure class=site-avatar><a href=/><img src=/img/avatar_hue825486955cd7c56d95e38b4bd2a8e3c_229979_300x0_resize_box_3.png width=300 height=300 class=site-logo loading=lazy alt=Avatar></a></figure><div class=site-meta><h1 class=site-name><a href=/>ZYChimne</a></h1><h2 class=site-description>Computer Science, Wuhan University</h2></div></header><ol class=social-menu><li><a href=https://github.com/ZYChimne target=_blank title=GitHub><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-brand-github" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M9 19c-4.3 1.4-4.3-2.5-6-3m12 5v-3.5c0-1 .1-1.4-.5-2 2.8-.3 5.5-1.4 5.5-6a4.6 4.6.0 00-1.3-3.2 4.2 4.2.0 00-.1-3.2s-1.1-.3-3.5 1.3a12.3 12.3.0 00-6.2.0C6.5 2.8 5.4 3.1 5.4 3.1a4.2 4.2.0 00-.1 3.2A4.6 4.6.0 004 9.5c0 4.6 2.7 5.7 5.5 6-.6.6-.6 1.2-.5 2V21"/></svg></a></li><li><a href=https://twitter.com/ZChimne target=_blank title=Twitter><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-brand-twitter" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"/><path d="M22 4.01c-1 .49-1.98.689-3 .99-1.121-1.265-2.783-1.335-4.38-.737S11.977 6.323 12 8v1c-3.245.083-6.135-1.395-8-4 0 0-4.182 7.433 4 11-1.872 1.247-3.739 2.088-6 2 3.308 1.803 6.913 2.423 10.034 1.517 3.58-1.04 6.522-3.723 7.651-7.742a13.84 13.84.0 00.497-3.753C20.18 7.773 21.692 5.25 22 4.009z"/></svg></a></li></ol><ol class=menu id=main-menu><li><a href=/><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-home" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><polyline points="5 12 3 12 12 3 21 12 19 12"/><path d="M5 12v7a2 2 0 002 2h10a2 2 0 002-2v-7"/><path d="M9 21v-6a2 2 0 012-2h2a2 2 0 012 2v6"/></svg><span>Home</span></a></li><li><a href=/about-me/><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-user" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="12" cy="7" r="4"/><path d="M6 21v-2a4 4 0 014-4h4a4 4 0 014 4v2"/></svg><span>About Me</span></a></li><div class=menu-bottom-section><li id=dark-mode-toggle><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-toggle-left" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="8" cy="12" r="2"/><rect x="2" y="6" width="20" height="12" rx="6"/></svg><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-toggle-right" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="16" cy="12" r="2"/><rect x="2" y="6" width="20" height="12" rx="6"/></svg><span>Dark Mode</span></li></div></ol></aside><main class="main full-width"><article class=main-article><header class=article-header><div class=article-details><div class=article-title-wrapper><h2 class=article-title><a href=/posts/2016-09-09-archaeal-richmond-mine-acidophilic/>Archaeal Richmond Mine Acidophilic Nanoorganisms</a></h2></div><footer class=article-time><div><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-calendar-time" width="56" height="56" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><path d="M11.795 21H5a2 2 0 01-2-2V7a2 2 0 012-2h12a2 2 0 012 2v4"/><circle cx="18" cy="18" r="4"/><path d="M15 3v4"/><path d="M7 3v4"/><path d="M3 11h16"/><path d="M18 16.496V18l1 1"/></svg><time class=article-time--published>Jan 13, 2020</time></div><div><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-clock" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="12" cy="12" r="9"/><polyline points="12 7 12 12 15 15"/></svg><time class=article-time--reading>6 minute read</time></div></footer></div></header><section class=article-content><p><strong>Archaeal Richmond Mine acidophilic nanoorganisms (ARMAN)</strong> were first discovered in an extremely acidic mine located in northern California (<a class=link href=https://en.wikipedia.org/wiki/Iron_Mountain_Mine title="Iron Mountain Mine" target=_blank rel=noopener>Richmond Mine at Iron Mountain</a>) by Brett Baker in Jill Banfield&rsquo;s laboratory at the <a class=link href=https://en.wikipedia.org/wiki/University_of_California_Berkeley title="University of California Berkeley" target=_blank rel=noopener>University of California Berkeley</a>. These novel groups of <a class=link href=https://en.wikipedia.org/wiki/Archaea title=Archaea target=_blank rel=noopener>archaea</a> named ARMAN-1, ARMAN-2 (<em><a class=link href=https://en.wikipedia.org/wiki/Candidatus title=Candidatus target=_blank rel=noopener>Candidatus</a> <a class=link href=https://en.wikipedia.org/wiki/Micrarchaeum_acidiphilum title="Micrarchaeum acidiphilum" target=_blank rel=noopener>Micrarchaeum acidiphilum</a></em> ARMAN-2 ), and ARMAN-3 were missed by previous PCR-based surveys of the mine community because the ARMANs have several mismatches with commonly used <a class=link href=https://en.wikipedia.org/wiki/PCR title=PCR target=_blank rel=noopener>PCR</a> primers for <a class=link href=https://en.wikipedia.org/wiki/16S_ribosomal_RNA title="16S ribosomal RNA" target=_blank rel=noopener>16S rRNA</a> genes. Baker <em>et al.</em><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-1 target=_blank rel=noopener>[1]</a> detected them in a later study using <a class=link href=https://en.wikipedia.org/wiki/Shotgun_sequencing title="Shotgun sequencing" target=_blank rel=noopener>shotgun sequencing</a> of the community. The three groups were originally thought to represent three unique lineages deeply branched within the <a class=link href=https://en.wikipedia.org/wiki/Euryarchaeota title=Euryarchaeota target=_blank rel=noopener>Euryarchaeota</a>, a subgroup of the <a class=link href=https://en.wikipedia.org/wiki/Archaea title=Archaea target=_blank rel=noopener>Archaea</a>. However, this has been revised, based on more complete archaeal genomic tree, that they belong to a super phylum named DPANN.<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-2 target=_blank rel=noopener>[2]</a> The ARMAN groups now comprise deeply divergent phyla named Micrarchaeota and Parvarchaeota.<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-3 target=_blank rel=noopener>[3]</a> Their 16S rRNA genes differ by as much as 17% between the three groups. Prior to their discovery all of the <a class=link href=https://en.wikipedia.org/wiki/Archaea title=Archaea target=_blank rel=noopener>Archaea</a> shown to be associated with Iron Mountain belonged to the order <a class=link href=https://en.wikipedia.org/wiki/Thermoplasmatales title=Thermoplasmatales target=_blank rel=noopener>Thermoplasmatales</a> (e.g., <em><a class=link href=https://en.wikipedia.org/wiki/Ferroplasma_acidarmanus title="Ferroplasma acidarmanus" target=_blank rel=noopener>Ferroplasma acidarmanus</a></em>).</p><h2 id=distribution>Distribution</h2><p>Examination of different sites in the mine using fluorescent probes specific to the ARMAN groups has revealed that they are always present in communities associated with <a class=link href=https://en.wikipedia.org/wiki/Acid_mine_drainage title="Acid mine drainage" target=_blank rel=noopener>acid mine drainage</a> (AMD), at Iron Mountain in northern California, that have pH &lt; 1.5. They are usually found in low abundance (5–25%) in the community. Recently, closely related organisms have been detected in an acidic boreal mire or <a class=link href=https://en.wikipedia.org/wiki/Bog title=Bog target=_blank rel=noopener>bog</a> in Finland,<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-4 target=_blank rel=noopener>[4]</a> another acid mine drainage site in extreme environments of <a class=link href=https://en.wikipedia.org/wiki/Rio_Tinto_%28river%29 title="Rio Tinto (river)" target=_blank rel=noopener>Rio Tinto</a>, southwestern Spain<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-5 target=_blank rel=noopener>[5]</a> and from weak alkaline deep subsurface hot spring in Yunohama, Japan.<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-6 target=_blank rel=noopener>[6]</a></p><h2 id=cell-structure-and-ecology>Cell structure and ecology</h2><p>Using <a class=link href=https://en.wikipedia.org/wiki/Cryo-electron_tomography title="Cryo-electron tomography" target=_blank rel=noopener>cryo-electron tomography</a> an extensive 3D characterization of uncultivated ARMAN cells within mine <a class=link href=https://en.wikipedia.org/wiki/Biofilms title=Biofilms target=_blank rel=noopener>biofilms</a> has been done (Comolli <em>et al.</em> 2009<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-7 target=_blank rel=noopener>[7]</a>). This has revealed that they are right at the cell size predicted to be the lower limit for life, 0.009 µm3 and 0.04 µm3 (NRC Steering group). They also found that despite their unusually small cell size it is common to find more than one type of <a class=link href=https://en.wikipedia.org/wiki/Virus title=Virus target=_blank rel=noopener>virus</a> attached to the cells while in the biofilms. Furthermore, the cells contain on average ≈92 <a class=link href=https://en.wikipedia.org/wiki/Ribosomes title=Ribosomes target=_blank rel=noopener>ribosomes</a> per cell, whereas the average <em><a class=link href=https://en.wikipedia.org/wiki/E._coli title="E. coli" target=_blank rel=noopener>E. coli</a></em> cell grown in culture contains ≈10,000 ribosomes. This suggests that for ARMAN cells a much more limited number of metabolites are present in a given cell. It raises questions about what the minimal requirements are for a living cell.</p><p>3D reconstructions of ARMAN cells in the environment has revealed that a small number of them attach to other Archaea of the order <a class=link href=https://en.wikipedia.org/wiki/Thermoplasmatales title=Thermoplasmatales target=_blank rel=noopener>Thermoplasmatales</a> (Baker <em>et al.</em> 2010 <a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-auto-8 target=_blank rel=noopener>[8]</a>). The Thermoplasmatales cells appear to penetrate the cell wall to the cytoplasm of the ARMAN cells.<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-9 target=_blank rel=noopener>[9]</a> The nature of this interaction hasn&rsquo;t been determined. It could be some sort of parasitic or symbiotic interaction. It is possible that ARMAN is getting some sort of metabolite that it is not able to produce on its own.</p><h2 id=genomics-and-proteomics>Genomics and proteomics</h2><p>The genomes of three ARMAN groups have been sequenced at the DOE <a class=link href=https://en.wikipedia.org/wiki/Joint_Genome_Institute title="Joint Genome Institute" target=_blank rel=noopener>Joint Genome Institute</a> during a 2006 <a class=link href=http://www.jgi.doe.gov/sequencing/why/3423.html target=_blank rel=noopener>Community Sequencing Program (CSP)</a>. These three genomes were successfully binned from the community genomic data using ESOM or Emergent <a class=link href=https://en.wikipedia.org/wiki/Self-Organizing_Map title="Self-Organizing Map" target=_blank rel=noopener>Self-Organizing Map</a> clustering of tetra-nucleotide DNA signatures (Dick <em>et al.</em> 2009<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-10 target=_blank rel=noopener>[10]</a>).</p><p>The first draft of <em>Candidatus</em> Micrarchaeum acidiphilum ARMAN-2 is ≈1 Mb (3 scaffolds, Baker <em>et al.</em>). PNAS 2010.<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-auto-8 target=_blank rel=noopener>[8]</a> The ARMAN-2 has recently been closed using 454 and Solexa sequencing of other biofilms to close the gaps and is being prepared for submission to NCBI. The genomes of ARMAN-4 and ARMAN-5 (roughly 1 Mb as well) have unusually average genes sizes, similar to those seen in endosymbiotic and parasitic bacteria. This may be signature of their inter-species interactions with other Archaea in nature (Baker <em>et al.</em> 2010 <a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-auto-8 target=_blank rel=noopener>[8]</a>). Furthermore, the branching of these groups near the <a class=link href="https://en.wikipedia.org/w/index.php?title=Euryarchaea&action=edit&redlink=1" title="Euryarchaea (page does not exist)" target=_blank rel=noopener>Euryarchaea</a>/<a class=link href=https://en.wikipedia.org/wiki/Crenarchaea title=Crenarchaea target=_blank rel=noopener>Crenarchaea</a> divide is reflected in them sharing many genetic aspects of both Crenarchaea and Euryarchaea. Specifically they have many genes that had previously only been identified in Crenarchaea. It is difficult to elucidate many of the commonly known metabolic pathways in ARMAN due to the unusually high number of unique genes that have been identified in their genomes.</p><p>A novel type of tRNA splicing endonuclease, involved in the processing of tRNA, has been discovered in ARMAN groups 1 and 2 (Fujishima <em>et al.</em> 2011<a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_note-11 target=_blank rel=noopener>[11]</a>). The enzyme consists of two duplicated catalytic units and one structural unit encoded on a single gene, representing a novel three-unit architecture.</p><h2 id=references>References</h2><ol><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-1 target=_blank rel=noopener>^</a></strong> Baker, Brett J.; et al. (2006). &ldquo;Lineages of Acidophilic Archaea Revealed by Community Genomic Analysis&rdquo;. <em><a class=link href=https://en.wikipedia.org/wiki/Science_%28journal%29 title="Science (journal)" target=_blank rel=noopener>Science</a></em>. <strong>314</strong> (5807): 1933–1935. <a class=link href=https://en.wikipedia.org/wiki/Bibcode title=Bibcode target=_blank rel=noopener>Bibcode</a>:<a class=link href=https://ui.adsabs.harvard.edu/abs/2006Sci...314.1933B target=_blank rel=noopener>2006Sci&mldr;314.1933B</a>. <a class=link href=https://en.wikipedia.org/wiki/Digital_object_identifier title="Digital object identifier" target=_blank rel=noopener>doi</a>:<a class=link href=https://doi.org/10.1126%2Fscience.1132690 target=_blank rel=noopener>10.1126/science.1132690</a>. <a class=link href=https://en.wikipedia.org/wiki/PubMed_Identifier title="PubMed Identifier" target=_blank rel=noopener>PMID</a> <a class=link href=https://pubmed.ncbi.nlm.nih.gov/17185602 target=_blank rel=noopener>17185602</a>.</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-2 target=_blank rel=noopener>^</a></strong> Rinke, C; Schwientek, P; Sczyrba, A; Ivanova, NN; Anderson, IJ; Cheng, JF; Darling, A; Malfatti, S; Swan, BK; Gies, EA; Dodsworth, JA; Hedlund, BP; Tsiamis, G; Sievert, SM; Liu, WT; Eisen, JA; Hallam, SJ; Kyrpides, NC; Stepanauskas, R; Rubin, EM; Hugenholtz, P; Woyke, T (2013). &ldquo;Insights into the phylogeny and coding potential of microbial dark matter&rdquo;. <em>Nature</em>. <strong>499</strong> (7459): 431–437. <a class=link href=https://en.wikipedia.org/wiki/Bibcode title=Bibcode target=_blank rel=noopener>Bibcode</a>:<a class=link href=https://ui.adsabs.harvard.edu/abs/2013Natur.499..431R target=_blank rel=noopener>2013Natur.499..431R</a>. <a class=link href=https://en.wikipedia.org/wiki/Digital_object_identifier title="Digital object identifier" target=_blank rel=noopener>doi</a>:<a class=link href=https://doi.org/10.1038%2Fnature12352 target=_blank rel=noopener>10.1038/nature12352</a>. <a class=link href=https://en.wikipedia.org/wiki/Handle_System title="Handle System" target=_blank rel=noopener>hdl</a>:<a class=link href=https://hdl.handle.net/1912%2F6194 target=_blank rel=noopener>1912/6194</a>. <a class=link href=https://en.wikipedia.org/wiki/PubMed_Identifier title="PubMed Identifier" target=_blank rel=noopener>PMID</a> <a class=link href=https://pubmed.ncbi.nlm.nih.gov/23851394 target=_blank rel=noopener>23851394</a>.</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-3 target=_blank rel=noopener>^</a></strong> Castelle, CJ; Wrighton, KC; Thomas, BC; Hug, LA; Brown, CT; Wilkins, MJ; Frischkorn, KR; Tringe, SG; Singh, A; Markillie, LM; Taylor, RC; Williams, KH; Banfield, JF (2015). &ldquo;Genomic Expansion of Domain Archaea Highlights Roles for Organisms from New Phyla in Anaerobic Carbon Cycling&rdquo;. <em>Current Biology</em>. <strong>25</strong> (6): 690–701. <a class=link href=https://en.wikipedia.org/wiki/Digital_object_identifier title="Digital object identifier" target=_blank rel=noopener>doi</a>:<a class=link href=https://doi.org/10.1016%2Fj.cub.2015.01.014 target=_blank rel=noopener>10.1016/j.cub.2015.01.014</a>. <a class=link href=https://en.wikipedia.org/wiki/PubMed_Identifier title="PubMed Identifier" target=_blank rel=noopener>PMID</a> <a class=link href=https://pubmed.ncbi.nlm.nih.gov/25702576 target=_blank rel=noopener>25702576</a>.</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-4 target=_blank rel=noopener>^</a></strong> Juottonen <em>et al.</em> <a class=link href=http://www.nature.com/ismej/journal/v2/n11/full/ismej200866a.html target=_blank rel=noopener>Seasonality of rDNA- and rRNA-derived archaeal communities and methanogenic potential in a boreal mire</a>, ISME Journal 24 July 2008</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-5 target=_blank rel=noopener>^</a></strong> Amaral-Zettler <em>et al.</em> <a class=link href=http://www.nature.com/ismej/journal/vaop/ncurrent/full/ismej2010101a.html target=_blank rel=noopener>Microbial community structure across the tree of life in the extreme Río Tinto</a>, ISME Journal 2010</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-6 target=_blank rel=noopener>^</a></strong> Murakami <em>et al.</em> <a class=link href=http://aem.asm.org/content/early/2011/12/02/AEM.06811-11 target=_blank rel=noopener>Metatranscriptomic analysis of microbes in an ocean-front deep subsurface hot spring reveals novel small RNAs and type-specific tRNA degradation</a>, Appl Environ Microbiol 2011</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-7 target=_blank rel=noopener>^</a></strong> Commolli, LR <em>et al.</em> <a class=link href=http://www.nature.com/ismej/journal/v3/n2/full/ismej200899a.html target=_blank rel=noopener>Three-dimensional analysis of the structure and ecology of a novel, ultra-small archaeon</a>, ISME Journal 3, 159–167.</li><li>^ <a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-auto_8-0 target=_blank rel=noopener><em><strong>a</strong></em></a> <a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-auto_8-1 target=_blank rel=noopener><em><strong>b</strong></em></a> <a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-auto_8-2 target=_blank rel=noopener><em><strong>c</strong></em></a> Baker; et al. (2010). <a class=link href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2889320 target=_blank rel=noopener>&ldquo;Enigmatic, ultrasmall, uncultivated Archaea&rdquo;</a>. <em>Proc. Natl. Acad. Sci</em>. <strong>107</strong> (19): 8806–8811. <a class=link href=https://en.wikipedia.org/wiki/Bibcode title=Bibcode target=_blank rel=noopener>Bibcode</a>:<a class=link href=https://ui.adsabs.harvard.edu/abs/2010PNAS..107.8806B target=_blank rel=noopener>2010PNAS..107.8806B</a>. <a class=link href=https://en.wikipedia.org/wiki/Digital_object_identifier title="Digital object identifier" target=_blank rel=noopener>doi</a>:<a class=link href=https://doi.org/10.1073%2Fpnas.0914470107 target=_blank rel=noopener>10.1073/pnas.0914470107</a>. <a class=link href=https://en.wikipedia.org/wiki/PubMed_Central title="PubMed Central" target=_blank rel=noopener>PMC</a> <a class=link href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2889320 target=_blank rel=noopener>2889320</a>. <a class=link href=https://en.wikipedia.org/wiki/PubMed_Identifier title="PubMed Identifier" target=_blank rel=noopener>PMID</a> <a class=link href=https://pubmed.ncbi.nlm.nih.gov/20421484 target=_blank rel=noopener>20421484</a>.</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-9 target=_blank rel=noopener>^</a></strong> Sanders, Robert (3 May 2010). <a class=link href=http://newscenter.berkeley.edu/2010/05/03/ARMAN_semifinal/ target=_blank rel=noopener>&ldquo;Weird, ultra-small microbes turn up in acidic mine drainage&rdquo;</a>.</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-10 target=_blank rel=noopener>^</a></strong> Dick <em>et al.</em> <a class=link href=http://genomebiology.com/2009/10/8/R85 target=_blank rel=noopener>Community-wide analysis of microbial genome sequence signatures.</a> Genome Biology 10:R85</li><li><strong><a class=link href=https://en.wikipedia.org/wiki/Archaeal_Richmond_Mine_acidophilic_nanoorganisms#cite_ref-11 target=_blank rel=noopener>^</a></strong> Fujishima; et al. (2011). <a class=link href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3239211 target=_blank rel=noopener>&ldquo;A novel three-unit tRNA splicing endonuclease found in ultrasmall Archaea possesses broad substrate specificity&rdquo;</a>. <em>Nucleic Acids Res</em>. <strong>39</strong> (22): 9695–704. <a class=link href=https://en.wikipedia.org/wiki/Digital_object_identifier title="Digital object identifier" target=_blank rel=noopener>doi</a>:<a class=link href=https://doi.org/10.1093%2Fnar%2Fgkr692 target=_blank rel=noopener>10.1093/nar/gkr692</a>. <a class=link href=https://en.wikipedia.org/wiki/PubMed_Central title="PubMed Central" target=_blank rel=noopener>PMC</a> <a class=link href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3239211 target=_blank rel=noopener>3239211</a>. <a class=link href=https://en.wikipedia.org/wiki/PubMed_Identifier title="PubMed Identifier" target=_blank rel=noopener>PMID</a> <a class=link href=https://pubmed.ncbi.nlm.nih.gov/21880595 target=_blank rel=noopener>21880595</a>.</li></ol><h2 id=external-links>External links</h2><p>Extant</p><p><a class=link href=https://en.wikipedia.org/wiki/Life title=Life target=_blank rel=noopener>life</a></p><p>phyla/divisions by domain</p><p>from Hacker News <a class=link href=https://ift.tt/30gLAjR target=_blank rel=noopener>https://ift.tt/30gLAjR</a></p></section><footer class=article-footer><section class=article-copyright><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-copyright" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><circle cx="12" cy="12" r="9"/><path d="M14.5 9a3.5 4 0 100 6"/></svg><span>Licensed under CC BY-NC-SA 4.0</span></section></footer></article><footer class=site-footer><section class=copyright>&copy;
2020 -
2022 ZYChimne</section><section class=powerby>Built with <a href=https://gohugo.io/ target=_blank rel=noopener>Hugo</a><br>Theme <b><a href=https://github.com/CaiJimmy/hugo-theme-stack target=_blank rel=noopener data-version=3.11.0>Stack</a></b> designed by <a href=https://jimmycai.com target=_blank rel=noopener>Jimmy</a></section></footer><div class=pswp tabindex=-1 role=dialog aria-hidden=true><div class=pswp__bg></div><div class=pswp__scroll-wrap><div class=pswp__container><div class=pswp__item></div><div class=pswp__item></div><div class=pswp__item></div></div><div class="pswp__ui pswp__ui--hidden"><div class=pswp__top-bar><div class=pswp__counter></div><button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
<button class="pswp__button pswp__button--share" title=Share></button>
<button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
<button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button><div class=pswp__preloader><div class=pswp__preloader__icn><div class=pswp__preloader__cut><div class=pswp__preloader__donut></div></div></div></div></div><div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap"><div class=pswp__share-tooltip></div></div><button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)"></button>
<button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)"></button><div class=pswp__caption><div class=pswp__caption__center></div></div></div></div></div><script src=https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.js integrity="sha256-ePwmChbbvXbsO02lbM3HoHbSHTHFAeChekF1xKJdleo=" crossorigin=anonymous defer></script><script src=https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe-ui-default.min.js integrity="sha256-UKkzOn/w1mBxRmLLGrSeyB4e1xbrp4xylgAWb3M42pU=" crossorigin=anonymous defer></script><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/default-skin/default-skin.css integrity="sha256-c0uckgykQ9v5k+IqViZOZKc47Jn7KQil4/MP3ySA3F8=" crossorigin=anonymous><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.css integrity="sha256-SBLU4vv6CA6lHsZ1XyTdhyjJxCjPif/TRkjnsyGAGnE=" crossorigin=anonymous></main><aside class="sidebar right-sidebar sticky"><section class="widget archives"><div class=widget-icon><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-hash" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z"/><line x1="5" y1="9" x2="19" y2="9"/><line x1="5" y1="15" x2="19" y2="15"/><line x1="11" y1="4" x2="7" y2="20"/><line x1="17" y1="4" x2="13" y2="20"/></svg></div><h2 class="widget-title section-title">Table of contents</h2><div class=widget--toc><nav id=TableOfContents><ol><li><a href=#distribution>Distribution</a></li><li><a href=#cell-structure-and-ecology>Cell structure and ecology</a></li><li><a href=#genomics-and-proteomics>Genomics and proteomics</a></li><li><a href=#references>References</a></li><li><a href=#external-links>External links</a></li></ol></nav></div></section></aside></div><script src=https://cdn.jsdelivr.net/npm/node-vibrant@3.1.5/dist/vibrant.min.js integrity="sha256-5NovOZc4iwiAWTYIFiIM7DxKUXKWvpVEuMEPLzcm5/g=" crossorigin=anonymous></script><script type=text/javascript src=/ts/main.js defer></script>
<script>(function(){const e=document.createElement("link");e.href="https://fonts.googleapis.com/css2?family=Lato:wght@300;400;700&display=swap",e.type="text/css",e.rel="stylesheet",document.head.appendChild(e)})()</script></body></html>